<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.553">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Daniel Kapitan">
<meta name="author" content="Melle Sieswerda">
<meta name="author" content="Andre Dekker">
<meta name="keywords" content="OMOP, OpenEHR, FHIR, secondary use, data platform, digital platform">

<title>Beyond the paradox of interoperability in open health data standards</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="index_files/libs/clipboard/clipboard.min.js"></script>
<script src="index_files/libs/quarto-html/quarto.js"></script>
<script src="index_files/libs/quarto-html/popper.min.js"></script>
<script src="index_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="index_files/libs/quarto-html/anchor.min.js"></script>
<link href="index_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="index_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="index_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="index_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="index_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul class="collapse">
  <li><a href="#looking-beyond-the-paradox-of-interoperability" id="toc-looking-beyond-the-paradox-of-interoperability" class="nav-link active" data-scroll-target="#looking-beyond-the-paradox-of-interoperability">Looking beyond the paradox of interoperability</a></li>
  <li><a href="#the-paradox-of-open-for-health-data-standards" id="toc-the-paradox-of-open-for-health-data-standards" class="nav-link" data-scroll-target="#the-paradox-of-open-for-health-data-standards">The paradox of open for health data standards</a></li>
  <li><a href="#standarization-of-health-data-for-federated-learning" id="toc-standarization-of-health-data-for-federated-learning" class="nav-link" data-scroll-target="#standarization-of-health-data-for-federated-learning">Standarization of health data for federated learning</a></li>
  <li><a href="#health-data-standards-in-low--and-middle-income-countries" id="toc-health-data-standards-in-low--and-middle-income-countries" class="nav-link" data-scroll-target="#health-data-standards-in-low--and-middle-income-countries">Health data standards in low- and middle income countries</a></li>
  <li><a href="#conclusion-and-outlook" id="toc-conclusion-and-outlook" class="nav-link" data-scroll-target="#conclusion-and-outlook">Conclusion and outlook</a></li>
  <li><a href="#bibliography" id="toc-bibliography" class="nav-link" data-scroll-target="#bibliography">References</a></li>
  </ul>
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="index.pdf"><i class="bi bi-file-pdf"></i>PDF (elsevier)</a></li></ul></div></nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Beyond the paradox of interoperability in open health data standards</h1>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Authors</div>
  <div class="quarto-title-meta-heading">Affiliations</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author">Daniel Kapitan <a href="mailto:daniel@kapitan.net" class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> <a href="https://orcid.org/0000-0001-8979-9194" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Dutch Hospital Data
          </p>
        <p class="affiliation">
            PharmAccess Foundation
          </p>
        <p class="affiliation">
            Eindhoven University of Technology
          </p>
      </div>
    <div class="quarto-title-meta-contents">
    <p class="author">Melle Sieswerda </p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Integral Cancer Registry Netherlands
          </p>
        <p class="affiliation">
            Maastricht University
          </p>
      </div>
    <div class="quarto-title-meta-contents">
    <p class="author">Andre Dekker </p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Maastricht University
          </p>
      </div>
  </div>

<div class="quarto-title-meta">

      
  
    
  </div>
  
<div>
  <div class="abstract">
    <div class="block-title">Abstract</div>
    <p>In response to the proposal of Tsafnat et al.&nbsp;to converge towards three open health data standards, we discuss two specific contexts, namely standardization of i) health data for federated learning, and ii) health data sharing in low- and middle income countries (LMICs). Based on our ongoing work in both areas, we provide a critical reflection on the proposed alignment of using OpenEHR, FHIR and OMOP as the default standard for the three domains of clinical care and administration, data exchange and longitudinal analysis, respectively. We find that for the two contexts considered there, this trichotomy does not do justice to details that are crucial in real-world implementations. This perspetive describes more specific design principles and implementation choices for these two types of health data sharing. In both case, we observe a strong convergence towards FHIR, while OMOP is used as an alternative or in combination with FHIR for federated learning.</p>
  </div>
</div>

<div>
  <div class="keywords">
    <div class="block-title">Keywords</div>
    <p>OMOP, OpenEHR, FHIR, secondary use, data platform, digital platform</p>
  </div>
</div>

</header>


<section id="looking-beyond-the-paradox-of-interoperability" class="level2">
<h2 class="anchored" data-anchor-id="looking-beyond-the-paradox-of-interoperability">Looking beyond the paradox of interoperability</h2>
<p>“A paradox of health care interoperability is the existence of a large number of standards exists with significant overlap among them,” say Tsafnat et al., followed by a call to actions towards the health informatics community to put effort into establishing convergence and preventing collision <span class="citation" data-cites="tsafnat2024converge">(<a href="#ref-tsafnat2024converge" role="doc-biblioref">Tsafnat et al., 2024</a>)</span>. To do so, they propose to converge on three open standards, namely i) OpenEHR for clinical care and administration; ii) Fast Health Interoperability Resources (FHIR) for data exhange and iii) Observational Medical Outcomes Partnership Common Data Model (OMOP) for longitudinal analysis. They argue that open data standards, backed by engaged communities, hold an advantage over proprietary ones and therefore should be chosen as the steppingstones towards achieving true interoperability.</p>
<p>While we support their high-level rationale and intention, we feel their proposed trichotomy does not do justice to details that are crucial in real-world implementations. This viewpoint provides a critical reflection on the their proposed framework in three parts. First, we reflect on salient differences between the three open standards from the perspective of the notion of openness of digital platforms <span class="citation" data-cites="dereuver2018digital">(<a href="#ref-dereuver2018digital" role="doc-biblioref">de Reuver et al., 2018</a>)</span>, data platforms <span class="citation" data-cites="dereuver2022openness">(<a href="#ref-dereuver2022openness" role="doc-biblioref">de Reuver et al., 2022</a>)</span> and the paradox of open <span class="citation" data-cites="keller2021paradox">(<a href="#ref-keller2021paradox" role="doc-biblioref">Keller and Tarkowski, 2021</a>)</span>. Subsequently, we present our findings in designing and implementing health data platforms in two specific contexts, namely i) platforms for federated learning on shared health data; and ii) health data platforms for low and middle income countries (LMICs). We conclude with …</p>
</section>
<section id="the-paradox-of-open-for-health-data-standards" class="level2">
<h2 class="anchored" data-anchor-id="the-paradox-of-open-for-health-data-standards">The paradox of open for health data standards</h2>
<p>Besides the paradox of interoperability put forward by Tsafnat et al., we argue that although open standards are a necessary, but not sufficient condition for convergence of health data standarization. We posit that open source implementations of components, libraries etc. constitute another necessary condition for establishing a flourishing health data platform and associated ecosystem. Research on digital platforms underline the importance of the platform openness, not only in term of open standards, but also in term of extensibility of the code base, availibility of complements to the core technical platform (in our case the data standard itself) and availibility of executable pieces of software <span class="citation" data-cites="dereuver2018digital">(<a href="#ref-dereuver2018digital" role="doc-biblioref">de Reuver et al., 2018</a>)</span>. Only when these aspects of digital platforms are fullfilled can we resonably expect that the platform will indeed be longlived.</p>
<p>In what they call the paradox of open, Keller and Tarkowsi argue that this conventional approach of open standards and open source flourish under two types of conditions <span class="citation" data-cites="keller2021paradox">(<a href="#ref-keller2021paradox" role="doc-biblioref">Keller and Tarkowski, 2021</a>)</span>. First, projects where many people contribute to the creation of a common resource have proven succesful. This is the story of Wikipedia, OpenStreetMap, Blender.org, and the countless free software projects that provide much of the internet’s infrastructure. Indeed, Tsafnat et al.&nbsp;have explicitly taken into account that “an engaged and vibrant community is a major advantage for the longevity of the data standards it uses,” which has informed their proposal to converge towards OMOP, FHIR en OpenEHR. However, the emphasis on open source implementations is somewhat overlooked. This point is only mentioned in passing and indirectly, when Tsafnat et al.&nbsp;reference work done by Reynolds and Wyatt who already argued in 2011 “… for the superiority of open source licensing to promote safer, more effective health care information systems. We claim that open source licensing in health care information systems is essential to rational procurement strategy” <span class="citation" data-cites="reynolds2011open">(<a href="#ref-reynolds2011open" role="doc-biblioref">Reynolds and Wyatt, 2011</a>)</span>. We believe that a realistic assessment of the current position of an open standard within the wider context of availability of complementary components and open source implementations is equally important when choosing which standard to adopt.</p>
<p>This point is related to the second condition put forward by Keller and Tarkoswki, namely that the conventional open approach has proven fruitful when “opening up” is the result of external incentives or requirements, rather than voluntary actions. This is the story of publicly-funded knowledge production like Open Access academic publications, cultural heritage collections in the Public Domain, Open Educational Resources (OER), and Open Government data. A canonical example in the birth of the GSM standard, which was mandated by European legislation.<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a> Reflecting on this perspective on openness, we observer a salient difference between FHIR vis-a-vis OpenEHR and OMOP, namely that the former is the only one that has been mandated (or at least strongly recommended) in some jurisdictions. In the US, the Office of the National Coordinator for Health Information Technology (ONC) and the Centers for Medicare and Medicaid Services (CMS) has introduced a steady stream of new regulations, criteria, and deadlines in Health IT that has resulted in significant adoption of FHIR <span class="citation" data-cites="firely2023fhir">(<a href="#ref-firely2023fhir" role="doc-biblioref">Firely, 2023</a>)</span>. In India, the open Health Claims Exchange protocol specification - which is based on FHIR - has been mandated by the Indian government as the standard for e-claims handling <span class="citation" data-cites="hcx">(<a href="#ref-hcx" role="doc-biblioref"><span>“<span>HCX Protocol</span> v0.9,”</span> 2023</a>)</span>. The African Union recommends all new implementations and digital health system improvements use FHIR as the primary mechanism for data exchange <span class="citation" data-cites="tilahun2023african">(<a href="#ref-tilahun2023african" role="doc-biblioref">Tilahun et al., 2023</a>)</span>, but doesn’t say anything about the use of, for example, OpenEHR for administrative systems of record.</p>
<p>These external incentives have resulted in a large boost in both commercial and open source development activities in the FHIR ecosystem. One such example is the speed with which the Bulk FHIR API has been defined and implemented in almost all major implementations <span class="citation" data-cites="mandl2020push jones2021landscape">(<a href="#ref-jones2021landscape" role="doc-biblioref">Jones et al., 2021</a>; <a href="#ref-mandl2020push" role="doc-biblioref">Mandl et al., 2020</a>)</span>. It has also led to more people voluntarily contributing to FHIR-related open source projects, which has resulted in a wide offering of FHIR components across major technology stacks (Java, Python, .NET), thereby strengthening the first condition. By comparison, the ecosystem of OMOP and OpenEHR is have not yet profited from external incentives to grow their community …</p>
<ul>
<li>smaller than that of FHIR</li>
<li>The majority of OMOP components are run by Observational Health Data Science and Informatics (OHDSI) …Say something that OMOP still has a relatively large ecosystem, with R libraries, pyOMOP etc.</li>
<li>The OpenEHR, however, seems subscritical. We are not judging the content and approach of OpenEHR, but there are just so few implementations.</li>
</ul>
<p>Hence, we stress that beyond the evaluating the intrinic structure of an open standard (which one is most suitable for a given domain) and the community that supports the standard, we need to take into account the wider ecosystem of open source implementations, complementary components etc. From this wider perspective on the whole ecosystem around a standard, FHIR stands out as having the diverse and rich ecosystem because it has been mandated in certain jurisdictions. This is relevant when comparing these standards in real-world implementations.</p>
</section>
<section id="standarization-of-health-data-for-federated-learning" class="level2">
<h2 class="anchored" data-anchor-id="standarization-of-health-data-for-federated-learning">Standarization of health data for federated learning</h2>
<p>The current fragmentation in health data is one of the major barriers towards leveraging the potential medical data for machine learning (ML). Without access to sufficient data, ML will be prevented from reaching its full potential and, ultimately, from making the transition from research to clinical practice. Data like this is hard to obtain, because health data is highly sensitive and its usage is tightly regulated.</p>
<p>Federated learning (FL) is a learning paradigm that aims to address these issues of data governance and privacy by training algorithms collaboratively without exchanging the data itself <span class="citation" data-cites="rieke2020future teo2024federated">(<a href="#ref-rieke2020future" role="doc-biblioref">Rieke et al., 2020</a>; <a href="#ref-teo2024federated" role="doc-biblioref">Teo et al., 2024</a>)</span>. Based on ongoing work with the PLUGIN healthcare consortium (<a href="https://plugin.healthcare">https://plugin.healthcare</a>, in Dutch) we have detailed an architecture for FL for secondary use. To this purpose, the the National Health Data Infrastructure agreements for research, policy and innovation for the Dutch healthcare sector which have been adopted at the beginning of 2024, has been taken as the starting point <span class="citation" data-cites="healthri2024agreements">(<a href="#ref-healthri2024agreements" role="doc-biblioref">Health-RI, 2024</a>)</span>. <a href="#fig-healthri-architecture" class="quarto-xref">Figure&nbsp;1</a> shows a high level overview of the platform, which comprises three areas (multiple use, applications and generic features) and a total of 26 functional components (for details please refer to <span class="citation" data-cites="healthri2024agreements">(<a href="#ref-healthri2024agreements" role="doc-biblioref">Health-RI, 2024</a>)</span>).</p>
<div id="fig-healthri-architecture" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-fig" id="fig-healthri-architecture-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: Reference architecture for the Dutch health data infrastructure for research and innovation <span class="citation" data-cites="healthri2024agreements">(<a href="#ref-healthri2024agreements" role="doc-biblioref">Health-RI, 2024</a>)</span>
</figcaption>
<div aria-describedby="fig-healthri-architecture-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="health-ri-architecture.png" class="img-fluid figure-img">
</div>
</figure>
</div>
<p>The principle of FL has been actively researched and developed within the Dutch healthcare community. The first proof-of-concept of the so-called Personal Health Train <span class="citation" data-cites="vansoest2018using">(<a href="#ref-vansoest2018using" role="doc-biblioref"><strong>vansoest2018using?</strong></a>)</span> ultimately led to the open source vantage6 platform <span class="citation" data-cites="smits2022improved">(<a href="#ref-smits2022improved" role="doc-biblioref"><strong>smits2022improved?</strong></a>)</span>. One of the prerequisites of this architecture is that organizations that participate in a federation of ‘data stations’ use the same common data model (CDM) to make the data Findable, Accessible, Interoperable and Resusable. These FAIR data stations comprise components 7, 8 and 9 in <a href="#fig-healthri-architecture" class="quarto-xref">Figure&nbsp;1</a>, i.e.&nbsp;the data, metadata and APIs, respectively, through which this the data station can be accessed and used. FHIR has been chosen as the common data model quite early in the development because of its practicality, provenance of RESTful APIs out of the box, support of many healthcare terminologies and flexibility through the profiling mechanims <span class="citation" data-cites="choudhury2020personal">(<a href="#ref-choudhury2020personal" role="doc-biblioref">Choudhury et al., 2020</a>)</span>.</p>
<p>This solution design has gained traction since then. The CODA platform, which aims to implement a FL infrastructure in Canada, compared OMOP and FHIR and chose the latter as it has been found to support more granular mappings required for analytics <span class="citation" data-cites="mullie2023coda">(<a href="#ref-mullie2023coda" role="doc-biblioref">Mullie et al., 2023</a>)</span>. Given that conceptually OMOP can be viewed as a strict subset of FHIR, hybrid solutions using OMOP and FHIR combined have also been reported, such as the German KETOS platform <span class="citation" data-cites="gruendner2019ketos">(<a href="#ref-gruendner2019ketos" role="doc-biblioref">Gruendner et al., 2019</a>)</span> and the preliminary findings from the European GenoMed4All project which aims to connect clinical and -omics data <span class="citation" data-cites="cremonesi2023need">(<a href="#ref-cremonesi2023need" role="doc-biblioref">Cremonesi et al., 2023</a>)</span>.</p>
<p>Using FHIR as the CDM in a FL architecture for persistent storage for longitudinal analysis may strike many as counterintuitive. The fact that some recent studies the practical designs of such platforms only mention OMOP only attests to this misconception <span class="citation" data-cites="wong2022applying">(<a href="#ref-wong2022applying" role="doc-biblioref"><strong>wong2022applying?</strong></a>)</span>.</p>
<ul>
<li><p>we don’t say you can’t use OMOP for data stations, in fact the following projects do that:</p>
<ul>
<li>OHDSI analytics <span class="citation" data-cites="khalid2021standardized">(<a href="#ref-khalid2021standardized" role="doc-biblioref">Khalid et al., 2021</a>)</span></li>
<li>… (look for more)</li>
</ul></li>
<li><p>this distinction become irrelevant as OMOP will have FHIR interfaces, see https://omoponfhir.org</p></li>
<li><p>however, FHIR is the CDM we choose because it is a superset, allowing for example also administrative and claims data or claims data can’t fit in OMOP</p></li>
<li><p>Principle of FHIR Profiles can be tied to principle of late binding: allow ingest of widely different sources, and gradually but more constraints and validations as you move closer to a specific use case. If machine learning is the primary objective for secondary use, we want to be able to cast a wider net of relevant data, rather than having very detailed data</p></li>
<li><p>Add more arguments here …</p></li>
</ul>
</section>
<section id="health-data-standards-in-low--and-middle-income-countries" class="level2">
<h2 class="anchored" data-anchor-id="health-data-standards-in-low--and-middle-income-countries">Health data standards in low- and middle income countries</h2>
<p>The OpenHIE framework <span class="citation" data-cites="openhie">(<a href="#ref-openhie" role="doc-biblioref"><span>“<span>OpenHIE Framework</span> v5.0,”</span> 2022</a>)</span> has been adopted by many sub-Saharan African countries <span class="citation" data-cites="mamuye2022health">(<a href="#ref-mamuye2022health" role="doc-biblioref">Mamuye et al., 2022</a>)</span> as the architectural blueprint for implementing nation-wide health information exchanges (HIE), including Nigeria <span class="citation" data-cites="dalhatu2023paper">(<a href="#ref-dalhatu2023paper" role="doc-biblioref">Dalhatu et al., 2023</a>)</span>, Kenya <span class="citation" data-cites="thaiya2021adoption">(<a href="#ref-thaiya2021adoption" role="doc-biblioref">Thaiya et al., 2021</a>)</span> and Tanzania <span class="citation" data-cites="nsaghurwe2021one">(<a href="#ref-nsaghurwe2021one" role="doc-biblioref">Nsaghurwe et al., 2021</a>)</span>. Conceptually, the OpenHIE framework constitutes a framework for an open digital platform, that mostly focuses on transactional exchange of data, that is, primary data sharing. Given the need to also enable secondary data sharing for academic research, real-world evidence studies etc., African countries have, as a matter of course, extended the framework to include “data &amp; analytics services” as an additional domain.</p>
<p>Based on our direct involvement in implementing and designing health data platforms in these countries, we observed that FHIR was practically the only viable solution. Although FHIR is originally intened for creating a longitudinal database, in fact open source FHIR implementations such as the HAPI FHIR server are the most widely used standards for realizing the so-called Shared Health Record within the OpenHIE architecture. not intended for</p>
<p>In relation to the conventional perspective on ‘openness’, which originally focused on open source and open standards as discussed above, has been superseded by “… conflicts about privacy, economic value extraction, the emergence of artificial intelligence, and the destabilizing effects of dominant platforms on (democratic) societies. Instead of access to information, the control of personal data has emerged in the age of platforms as the critical contention” <span class="citation" data-cites="keller2021paradox">(<a href="#ref-keller2021paradox" role="doc-biblioref">Keller and Tarkowski, 2021</a>)</span>. These conflicts are particularly salient in the healthcare domain, where people are generally willing to share their health data to receive the best care (primary use), while the attitude towards secondary use of health data varies greatly depending on the type and context <span class="citation" data-cites="cascini2024health">(<a href="#ref-cascini2024health" role="doc-biblioref">Cascini et al., 2024</a>)</span>. Case studies on digital platforms in healthcare point to an emerging pattern where the focus shifts from the digital platform with its defining software and hardware components, to the data as the primary object of interest in and of itself <span class="citation" data-cites="ozalp2022digital alaimo2022organizations">(<a href="#ref-alaimo2022organizations" role="doc-biblioref">Alaimo and Kallinikos, 2022</a>; <a href="#ref-ozalp2022digital" role="doc-biblioref">Ozalp et al., 2022</a>)</span>. This observation ties into with the proposed research agenda by de Reuver et al.&nbsp;to consider data platforms as a phenomenon distinct from digital platforms <span class="citation" data-cites="dereuver2018digital dereuver2022openness">(<a href="#ref-dereuver2022openness" role="doc-biblioref">de Reuver et al., 2022</a>; <a href="#ref-dereuver2018digital" role="doc-biblioref">de Reuver et al., 2018</a>)</span>.</p>
<p>Essentially, “… this paradox is that openness of data is both a challenge to and an enabler of concentrations of power. The ideas of open access and free reuse of information goods continue to be some of the most powerful challenges to the exclusive control by corporations and states over information goods. Yet making such resources open also exposes them to the imbalances of power that shape these societies – and in the worst cases serves to strengthen these imbalances.” Put differently, having open standards is no guarantee that we will be able to breakthrough the current status quo with huge asymmetries in terms of bargaining power, access control and development resources related to health data sharing. We posit that having open data standards is a necessary condition in improving the current state of affairs, yet it is not a sufficient condition.</p>
<p>The shift in perspective from digital platforms supporting primary data sharing toward data platforms supporting secondary data sharing is a contentious issue both in high income countries, with for example the ongoing efforts to establish a European Health Data Space <span class="citation" data-cites="otto2022designing">(<a href="#ref-otto2022designing" role="doc-biblioref">Otto et al., 2022</a>)</span>, and low- and middle income countries that are aiming to deploy nationwide health information exchanges to support primary and secondary data use at once <span class="citation" data-cites="mamuye2022health">(<a href="#ref-mamuye2022health" role="doc-biblioref">Mamuye et al., 2022</a>)</span>. A better understanding of openness is particularly relevant if we are to realize a solidarity-based approach to health data sharing that i) gives people a greater control over their data as active decision makers; ii) ensures that the value of data is harnessed for public good; and iii) moves society towards equity and justice by counteracting dynamics of data extraction <span class="citation" data-cites="kickbusch2021lancet prainsack2022data prainsack2023beyond">(<a href="#ref-kickbusch2021lancet" role="doc-biblioref">Kickbusch et al., 2021</a>; <a href="#ref-prainsack2022data" role="doc-biblioref">Prainsack et al., 2022</a>; <a href="#ref-prainsack2023beyond" role="doc-biblioref">Prainsack and El-Sayed, 2023</a>)</span>.</p>
<p>More than a decade later, we observe that only a very small fraction of health IT systems are based on open source, the majority of which are used in LMICs which we will discuss later <span class="citation" data-cites="digitalpublicgoods">(<a href="#ref-digitalpublicgoods" role="doc-biblioref"><span>“Digital <span>Public Goods Alliance</span>,”</span> 2024</a>)</span>.</p>
</section>
<section id="conclusion-and-outlook" class="level2">
<h2 class="anchored" data-anchor-id="conclusion-and-outlook">Conclusion and outlook</h2>
<ul>
<li>We underline the need for open data standards as a necessary condition to achieve interoperability</li>
<li>It is not a sufficient condition</li>
<li>MPC as next step up from FL</li>
</ul>
</section>
<section id="bibliography" class="level1 unnumbered">


</section>


<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-alaimo2022organizations" class="csl-entry" role="listitem">
Alaimo, C., Kallinikos, J., 2022. Organizations <span>Decentered</span>: <span>Data Objects</span>, <span>Technology</span> and <span>Knowledge</span>. Organization Science 33, 19–37. <a href="https://doi.org/10.1287/orsc.2021.1552">https://doi.org/10.1287/orsc.2021.1552</a>
</div>
<div id="ref-cascini2024health" class="csl-entry" role="listitem">
Cascini, F., Pantovic, A., Al-Ajlouni, Y.A., Puleo, V., De Maio, L., Ricciardi, W., 2024. Health data sharing attitudes towards primary and secondary use of data: A systematic review. eClinicalMedicine 71, 102551. <a href="https://doi.org/10.1016/j.eclinm.2024.102551">https://doi.org/10.1016/j.eclinm.2024.102551</a>
</div>
<div id="ref-choudhury2020personal" class="csl-entry" role="listitem">
Choudhury, A., van Soest, J., Nayak, S., Dekker, A., 2020. Personal <span>Health Train</span> on <span>FHIR</span>: <span>A Privacy Preserving Federated Approach</span> for <span>Analyzing FAIR Data</span> in <span>Healthcare</span>, in: Bhattacharjee, A., Borgohain, S.Kr., Soni, B., Verma, G., Gao, X.-Z. (Eds.), Machine <span>Learning</span>, <span>Image Processing</span>, <span>Network Security</span> and <span>Data Sciences</span>, Communications in <span>Computer</span> and <span>Information Science</span>. Springer, Singapore, pp. 85–95. <a href="https://doi.org/10.1007/978-981-15-6315-7_7">https://doi.org/10.1007/978-981-15-6315-7_7</a>
</div>
<div id="ref-cremonesi2023need" class="csl-entry" role="listitem">
Cremonesi, F., Planat, V., Kalokyri, V., Kondylakis, H., Sanavia, T., Miguel Mateos Resinas, V., Singh, B., Uribe, S., 2023. The need for multimodal health data modeling: <span>A</span> practical approach for a federated-learning healthcare platform. Journal of Biomedical Informatics 141, 104338. <a href="https://doi.org/10.1016/j.jbi.2023.104338">https://doi.org/10.1016/j.jbi.2023.104338</a>
</div>
<div id="ref-dalhatu2023paper" class="csl-entry" role="listitem">
Dalhatu, I., Aniekwe, C., Bashorun, A., Abdulkadir, A., Dirlikov, E., Ohakanu, S., Adedokun, O., Oladipo, A., Jahun, I., Murie, L., Yoon, S., Abdu-Aguye, M.G., Sylvanus, A., Indyer, S., Abbas, I., Bello, M., Nalda, N., Alagi, M., Odafe, S., Adebajo, S., Ogorry, O., Akpu, M., Okoye, I., Kakanfo, K., Onovo, A.A., Ashefor, G., Nzelu, C., Ikpeazu, A., Aliyu, G., Ellerbrock, T., Boyd, M., Stafford, K.A., Swaminathan, M., 2023. From <span>Paper Files</span> to <span>Web-Based Application</span> for <span>Data-Driven Monitoring</span> of <span>HIV Programs</span>: <span>Nigeria</span>’s <span>Journey</span> to a <span>National Data Repository</span> for <span>Decision-Making</span> and <span>Patient Care</span>. Methods of Information in Medicine 62, 130–139. <a href="https://doi.org/10.1055/s-0043-1768711">https://doi.org/10.1055/s-0043-1768711</a>
</div>
<div id="ref-dereuver2022openness" class="csl-entry" role="listitem">
de Reuver, M., Ofe, H., Agahari, W., Abbas, A.E., Zuiderwijk, A., 2022. The openness of data platforms: A research agenda, in: Proceedings of the 1st <span>International Workshop</span> on <span>Data Economy</span>, <span>DE</span> ’22. Association for Computing Machinery, New York, NY, USA, pp. 34–41. <a href="https://doi.org/10.1145/3565011.3569056">https://doi.org/10.1145/3565011.3569056</a>
</div>
<div id="ref-dereuver2018digital" class="csl-entry" role="listitem">
de Reuver, M., Sørensen, C., Basole, R.C., 2018. The <span>Digital Platform</span>: <span>A Research Agenda</span>. Journal of Information Technology 33, 124–135. <a href="https://doi.org/10.1057/s41265-016-0033-3">https://doi.org/10.1057/s41265-016-0033-3</a>
</div>
<div id="ref-digitalpublicgoods" class="csl-entry" role="listitem">
Digital <span>Public Goods Alliance</span>, 2024. Digital Public Goods Alliance - Promoting digital public goods to create a more equitable world.
</div>
<div id="ref-firely2023fhir" class="csl-entry" role="listitem">
Firely, 2023. <span>FHIR</span> in <span>US</span> healthcare regulations.
</div>
<div id="ref-gruendner2019ketos" class="csl-entry" role="listitem">
Gruendner, J., Schwachhofer, T., Sippl, P., Wolf, N., Erpenbeck, M., Gulden, C., Kapsner, L.A., Zierk, J., Mate, S., Stürzl, M., Croner, R., Prokosch, H.-U., Toddenroth, D., 2019. <span>KETOS</span>: <span>Clinical</span> decision support and machine learning as a service – <span>A</span> training and deployment platform based on <span>Docker</span>, <span>OMOP-CDM</span>, and <span>FHIR Web Services</span>. PLOS ONE 14, e0223010. <a href="https://doi.org/10.1371/journal.pone.0223010">https://doi.org/10.1371/journal.pone.0223010</a>
</div>
<div id="ref-hcx" class="csl-entry" role="listitem">
<span>HCX Protocol</span> v0.9, 2023.
</div>
<div id="ref-healthri2024agreements" class="csl-entry" role="listitem">
Health-RI, 2024. Agreements on the <span>National Health Data Infrastructure</span> for <span>Research</span>, <span>Policy</span> and <span>Innovation</span> - <span class="nocase">Health-RI Nationale Gezondheidsdata-infrastructuur</span> - <span>Confluence</span>.
</div>
<div id="ref-jones2021landscape" class="csl-entry" role="listitem">
Jones, J., Gottlieb, D., Mandel, J.C., Ignatov, V., Ellis, A., Kubick, W., Mandl, K.D., 2021. A landscape survey of planned <span>SMART</span>/<span>HL7</span> bulk <span>FHIR</span> data access <span>API</span> implementations and tools. Journal of the American Medical Informatics Association 28, 1284–1287. <a href="https://doi.org/10.1093/jamia/ocab028">https://doi.org/10.1093/jamia/ocab028</a>
</div>
<div id="ref-keller2021paradox" class="csl-entry" role="listitem">
Keller, P., Tarkowski, A., 2021. The <span>Paradox</span> of <span>Open</span>. Open Future.
</div>
<div id="ref-khalid2021standardized" class="csl-entry" role="listitem">
Khalid, S., Yang, C., Blacketer, C., Duarte-Salles, T., Fernández-Bertolín, S., Kim, C., Park, R.W., Park, J., Schuemie, M.J., Sena, A.G., Suchard, M.A., You, S.C., Rijnbeek, P.R., Reps, J.M., 2021. A standardized analytics pipeline for reliable and rapid development and validation of prediction models using observational health data. Computer Methods and Programs in Biomedicine 211, 106394. <a href="https://doi.org/10.1016/j.cmpb.2021.106394">https://doi.org/10.1016/j.cmpb.2021.106394</a>
</div>
<div id="ref-kickbusch2021lancet" class="csl-entry" role="listitem">
Kickbusch, I., Piselli, D., Agrawal, A., Balicer, R., Banner, O., Adelhardt, M., Capobianco, E., Fabian, C., Singh Gill, A., Lupton, D., Medhora, R.P., Ndili, N., Ryś, A., Sambuli, N., Settle, D., Swaminathan, S., Morales, J.V., Wolpert, M., Wyckoff, A.W., Xue, L., Bytyqi, A., Franz, C., Gray, W., Holly, L., Neumann, M., Panda, L., Smith, R.D., Georges Stevens, E.A., Wong, B.L.H., 2021. The <span>Lancet</span> and <span>Financial Times Commission</span> on governing health futures 2030: Growing up in a digital world. The Lancet 398, 1727–1776. <a href="https://doi.org/10.1016/S0140-6736(21)01824-9">https://doi.org/10.1016/S0140-6736(21)01824-9</a>
</div>
<div id="ref-mamuye2022health" class="csl-entry" role="listitem">
Mamuye, A.L., Yilma, T.M., Abdulwahab, A., Broomhead, S., Zondo, P., Kyeng, M., Maeda, J., Abdulaziz, M., Wuhib, T., Tilahun, B.C., 2022. Health information exchange policy and standards for digital health systems in africa: <span>A</span> systematic review. PLOS Digital Health 1, e0000118. <a href="https://doi.org/10.1371/journal.pdig.0000118">https://doi.org/10.1371/journal.pdig.0000118</a>
</div>
<div id="ref-mandl2020push" class="csl-entry" role="listitem">
Mandl, K.D., Gottlieb, D., Mandel, J.C., Ignatov, V., Sayeed, R., Grieve, G., Jones, J., Ellis, A., Culbertson, A., 2020. Push <span>Button Population Health</span>: <span>The SMART</span>/<span>HL7 FHIR Bulk Data Access Application Programming Interface</span>. npj Digital Medicine 3, 1–9. <a href="https://doi.org/10.1038/s41746-020-00358-4">https://doi.org/10.1038/s41746-020-00358-4</a>
</div>
<div id="ref-mullie2023coda" class="csl-entry" role="listitem">
Mullie, L., Afilalo, J., Archambault, P., Bouchakri, R., Brown, K., Buckeridge, D.L., Cavayas, Y.A., Turgeon, A.F., Martineau, D., Lamontagne, F., Lebrasseur, M., Lemieux, R., Li, J., Sauthier, M., St-Onge, P., Tang, A., Witteman, W., Chassé, M., 2023. <span>CODA</span>: An open-source platform for federated analysis and machine learning on distributed healthcare data. Journal of the American Medical Informatics Association ocad235. <a href="https://doi.org/10.1093/jamia/ocad235">https://doi.org/10.1093/jamia/ocad235</a>
</div>
<div id="ref-nsaghurwe2021one" class="csl-entry" role="listitem">
Nsaghurwe, A., Dwivedi, V., Ndesanjo, W., Bamsi, H., Busiga, M., Nyella, E., Massawe, J.V., Smith, D., Onyejekwe, K., Metzger, J., Taylor, P., 2021. One country’s journey to interoperability: <span>Tanzania</span>’s experience developing and implementing a national health information exchange. BMC Medical Informatics and Decision Making 21, 139. <a href="https://doi.org/10.1186/s12911-021-01499-6">https://doi.org/10.1186/s12911-021-01499-6</a>
</div>
<div id="ref-openhie" class="csl-entry" role="listitem">
<span>OpenHIE Framework</span> v5.0, 2022. OpenHIE.
</div>
<div id="ref-otto2022designing" class="csl-entry" role="listitem">
Otto, B., Ten Hompel, M., Wrobel, S. (Eds.), 2022. Designing <span>Data Spaces</span>: <span>The Ecosystem Approach</span> to <span>Competitive Advantage</span>. Springer International Publishing, Cham. <a href="https://doi.org/10.1007/978-3-030-93975-5">https://doi.org/10.1007/978-3-030-93975-5</a>
</div>
<div id="ref-ozalp2022digital" class="csl-entry" role="listitem">
Ozalp, H., Ozcan, P., Dinckol, D., Zachariadis, M., Gawer, A., 2022. <span>“<span>Digital Colonization</span>”</span> of <span>Highly Regulated Industries</span>: <span>An Analysis</span> of <span>Big Tech Platforms</span>’ <span>Entry</span> into <span>Health Care</span> and <span>Education</span>. California Management Review 64, 78–107. <a href="https://doi.org/10.1177/00081256221094307">https://doi.org/10.1177/00081256221094307</a>
</div>
<div id="ref-prainsack2023beyond" class="csl-entry" role="listitem">
Prainsack, B., El-Sayed, S., 2023. Beyond <span>Individual Rights</span>: <span>How Data Solidarity Gives People Meaningful Control</span> over <span>Data</span>. The American Journal of Bioethics 23, 36–39. <a href="https://doi.org/10.1080/15265161.2023.2256267">https://doi.org/10.1080/15265161.2023.2256267</a>
</div>
<div id="ref-prainsack2022data" class="csl-entry" role="listitem">
Prainsack, B., El-Sayed, S., Forgó, N., Szoszkiewicz, Ł., Baumer, P., 2022. Data solidarity: A blueprint for governing health futures. The Lancet Digital Health 4, e773–e774. <a href="https://doi.org/10.1016/S2589-7500(22)00189-3">https://doi.org/10.1016/S2589-7500(22)00189-3</a>
</div>
<div id="ref-reynolds2011open" class="csl-entry" role="listitem">
Reynolds, C.J., Wyatt, J.C., 2011. Open <span>Source</span>, <span>Open Standards</span>, and <span>Health Care Information Systems</span>. Journal of Medical Internet Research 13, e1521. <a href="https://doi.org/10.2196/jmir.1521">https://doi.org/10.2196/jmir.1521</a>
</div>
<div id="ref-rieke2020future" class="csl-entry" role="listitem">
Rieke, N., Hancox, J., Li, W., Milletarì, F., Roth, H.R., Albarqouni, S., Bakas, S., Galtier, M.N., Landman, B.A., Maier-Hein, K., Ourselin, S., Sheller, M., Summers, R.M., Trask, A., Xu, D., Baust, M., Cardoso, M.J., 2020. The future of digital health with federated learning. npj Digital Medicine 3, 1–7. <a href="https://doi.org/10.1038/s41746-020-00323-1">https://doi.org/10.1038/s41746-020-00323-1</a>
</div>
<div id="ref-teo2024federated" class="csl-entry" role="listitem">
Teo, Z.L., Jin, L., Liu, N., Li, S., Miao, D., Zhang, X., Ng, W.Y., Tan, T.F., Lee, D.M., Chua, K.J., Heng, J., Liu, Y., Goh, R.S.M., Ting, D.S.W., 2024. Federated machine learning in healthcare: <span>A</span> systematic review on clinical applications and technical architecture. Cell Reports Medicine 5, 101419. <a href="https://doi.org/10.1016/j.xcrm.2024.101419">https://doi.org/10.1016/j.xcrm.2024.101419</a>
</div>
<div id="ref-thaiya2021adoption" class="csl-entry" role="listitem">
Thaiya, M.S., Julia, K., Joram, M., Benard, M., Nambiro, D.A., 2021. Adoption of <span>ICT</span> to <span>Enhance Access</span> to <span>Healthcare</span> in <span>Kenya</span>.
</div>
<div id="ref-tilahun2023african" class="csl-entry" role="listitem">
Tilahun, B., Mamuye, A., Yilma, T., Shehata, Y., 2023. African <span>Union Health Information Exchange Guidelines</span> and <span>Standards</span>.
</div>
<div id="ref-tsafnat2024converge" class="csl-entry" role="listitem">
Tsafnat, G., Dunscombe, R., Gabriel, D., Grieve, G., Reich, C., 2024. Converge or collide? <span>Making</span> sense of a plethora of open data standards in healthcare: An editorial.
</div>
</div></section><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>See <a href="https://en.wikipedia.org/wiki/GSM#Initial_European_development">https://en.wikipedia.org/wiki/GSM#Initial_European_development</a> for details.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>