<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.40">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Daniel Kapitan">
<meta name="author" content="Femke Heddema">
<meta name="author" content="Andre Dekker">
<meta name="author" content="Melle Sieswerda">
<meta name="author" content="Bart-Jan Verhoeff">
<meta name="author" content="Matt Berg">
<meta name="keywords" content="health data interoperability, OMOP, openEHR, FHIR, secondary use, data platform, digital platform">

<title>Data interoperability in context: the importance of open-source implementations when choosing open standards</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="index_files/libs/clipboard/clipboard.min.js"></script>
<script src="index_files/libs/quarto-html/quarto.js"></script>
<script src="index_files/libs/quarto-html/popper.min.js"></script>
<script src="index_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="index_files/libs/quarto-html/anchor.min.js"></script>
<link href="index_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="index_files/libs/quarto-html/quarto-syntax-highlighting-549806ee2085284f45b00abea8c6df48.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="index_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="index_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="index_files/libs/bootstrap/bootstrap-973236bd072d72a04ee9cd82dcc9cb29.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul class="collapse">
  <li><a href="#open-standards-are-a-necessary-but-not-sufficient-condition-for-interoperability" id="toc-open-standards-are-a-necessary-but-not-sufficient-condition-for-interoperability" class="nav-link active" data-scroll-target="#open-standards-are-a-necessary-but-not-sufficient-condition-for-interoperability">Open standards are a necessary but not sufficient condition for interoperability</a></li>
  <li><a href="#digital-platforms-require-extensibility-availability-of-complementary-components-and-availability-of-executable-pieces-of-software" id="toc-digital-platforms-require-extensibility-availability-of-complementary-components-and-availability-of-executable-pieces-of-software" class="nav-link" data-scroll-target="#digital-platforms-require-extensibility-availability-of-complementary-components-and-availability-of-executable-pieces-of-software">Digital platforms require extensibility, availability of complementary components and availability of executable pieces of software</a></li>
  <li><a href="#standardization-of-health-data-for-federated-learning" id="toc-standardization-of-health-data-for-federated-learning" class="nav-link" data-scroll-target="#standardization-of-health-data-for-federated-learning">Standardization of health data for federated learning</a></li>
  <li><a href="#health-data-standards-in-lmics" id="toc-health-data-standards-in-lmics" class="nav-link" data-scroll-target="#health-data-standards-in-lmics">Health data standards in LMICs</a></li>
  <li><a href="#discussion-conclusion-and-future-research" id="toc-discussion-conclusion-and-future-research" class="nav-link" data-scroll-target="#discussion-conclusion-and-future-research">Discussion, conclusion and future research</a></li>
  <li><a href="#authors-contributions" id="toc-authors-contributions" class="nav-link" data-scroll-target="#authors-contributions">Authors’ Contributions</a></li>
  <li><a href="#acknowledgements" id="toc-acknowledgements" class="nav-link" data-scroll-target="#acknowledgements">Acknowledgements</a></li>
  <li><a href="#conflicts-of-interests" id="toc-conflicts-of-interests" class="nav-link" data-scroll-target="#conflicts-of-interests">Conflicts of interests</a></li>
  <li><a href="#abbreviations" id="toc-abbreviations" class="nav-link" data-scroll-target="#abbreviations">Abbreviations</a></li>
  <li><a href="#bibliography" id="toc-bibliography" class="nav-link" data-scroll-target="#bibliography">References</a></li>
  </ul>
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="index.pdf"><i class="bi bi-file-pdf"></i>PDF</a></li><li><a href="index.docx"><i class="bi bi-file-word"></i>MS Word</a></li></ul></div></nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Data interoperability in context: the importance of open-source implementations when choosing open standards</h1>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Authors</div>
  <div class="quarto-title-meta-heading">Affiliations</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author">Daniel Kapitan <a href="mailto:daniel@kapitan.net" class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> <a href="https://orcid.org/0000-0001-8979-9194" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Dutch Hospital Data
          </p>
        <p class="affiliation">
            PharmAccess Foundation
          </p>
        <p class="affiliation">
            Eindhoven University of Technology
          </p>
      </div>
    <div class="quarto-title-meta-contents">
    <p class="author">Femke Heddema <a href="https://orcid.org/0009-0007-4240-9786" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            PharmAccess Foundation
          </p>
      </div>
    <div class="quarto-title-meta-contents">
    <p class="author">Andre Dekker <a href="https://orcid.org/0000-0002-0422-7996" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Department of Radiation Oncology (Maastro), GROW - Research Institute for Oncology and Reproduction, Maastricht University Medical Center+
          </p>
      </div>
    <div class="quarto-title-meta-contents">
    <p class="author">Melle Sieswerda <a href="https://orcid.org/0000-0003-2025-0771" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Integral Cancer Registry Netherlands
          </p>
        <p class="affiliation">
            Maastricht University
          </p>
      </div>
    <div class="quarto-title-meta-contents">
    <p class="author">Bart-Jan Verhoeff </p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Expertisecentrum Zorgalgoritmen
          </p>
      </div>
    <div class="quarto-title-meta-contents">
    <p class="author">Matt Berg </p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Ona
          </p>
      </div>
  </div>

<div class="quarto-title-meta">

      
  
    
  </div>
  
<div>
  <div class="abstract">
    <div class="block-title">Abstract</div>
    <p>In response to the proposal of Tsafnat et al.&nbsp;to converge towards three open health data standards, this viewpoint provides a critical reflection on the proposed alignment of using openEHR, FHIR and OMOP as the default standards for clinical care and administration, data exchange and longitudinal analysis, respectively. We argue that open standards are a necessary but not sufficient condition to achieve health data interoperability. The ecosystem of open-source software needs to be considered when choosing an appropriate standard for a given context. We discuss two specific contexts, namely standardization of i) health data for federated learning, and ii) health data sharing in low and middle income countries (LMICs). Specific design principles, practical considerations and implementation choices for these two contexts are described, based on ongoing work in both areas. In the case of federated learning, we observe convergence towards OMOP and FHIR, where the two standards can effectively be used side-by-side given the availability of mediators between the two. In the case of health information exchanges in LMICs, we see a strong convergence towards FHIR as the primary standard. We propose practical guidelines for context-specific adaptation of open health data standards.</p>
  </div>
</div>

<div>
  <div class="keywords">
    <div class="block-title">Keywords</div>
    <p>health data interoperability, OMOP, openEHR, FHIR, secondary use, data platform, digital platform</p>
  </div>
</div>

</header>


<section id="open-standards-are-a-necessary-but-not-sufficient-condition-for-interoperability" class="level2">
<h2 class="anchored" data-anchor-id="open-standards-are-a-necessary-but-not-sufficient-condition-for-interoperability">Open standards are a necessary but not sufficient condition for interoperability</h2>
<p>“A paradox of health care interoperability is the existence of a large number of standards with significant overlap among them,” say Tsafnat et al., followed by a call to action towards the health informatics community to put effort into establishing convergence and preventing collision <span class="citation" data-cites="tsafnat2024converge">[<a href="#ref-tsafnat2024converge" role="doc-biblioref">1</a>]</span>. To do so, they propose to converge on three open standards, namely i) openEHR for clinical care and administration; ii) Fast Health Interoperability Resources (FHIR) for data exchange and iii) Observational Medical Outcomes Partnership Common Data Model (OMOP) for longitudinal analysis. They argue that open data standards, backed by engaged communities, hold an advantage over proprietary ones and therefore should be chosen as the steppingstones towards achieving true interoperability.</p>
<p>While we support their high-level rationale and intention, we feel their proposed trichotomy does not do justice to details that are crucial in real-world implementations. This viewpoint provides a critical reflection on their proposed framework in three parts. First, we reflect on salient differences between the three open standards from the perspective of the notion of openness of digital platforms <span class="citation" data-cites="dereuver2018digital">[<a href="#ref-dereuver2018digital" role="doc-biblioref">2</a>]</span>, the paradox of open <span class="citation" data-cites="keller2021paradox">[<a href="#ref-keller2021paradox" role="doc-biblioref">3</a>]</span> and the hourglass model of open architectures <span class="citation" data-cites="estrin2010health beck2019hourglass">[<a href="#ref-estrin2010health" role="doc-biblioref">4</a>,<a href="#ref-beck2019hourglass" role="doc-biblioref">5</a>]</span>. Subsequently, we outline the importance of the open-source software (OSS) by reflecting on our considerations in designing and implementing health data platforms in two specific contexts, namely i) platforms for federated learning on shared health data in high income countries; and ii) health data platforms for low and middle income countries (LMICs). These case studies illustrate the limitations of the trichotomy proposed by Tsafnat et. Particularly, we argue that of the three standards, FHIR stands out as being the most practical and adaptable which allows it to be used for longitudinal analysis and routine collection of clinical data as well, besides its original purpose as a health data exchange standard. We conclude this viewpoint with practical implications of these findings and directions for future research of open health data standards.</p>
</section>
<section id="digital-platforms-require-extensibility-availability-of-complementary-components-and-availability-of-executable-pieces-of-software" class="level2">
<h2 class="anchored" data-anchor-id="digital-platforms-require-extensibility-availability-of-complementary-components-and-availability-of-executable-pieces-of-software">Digital platforms require extensibility, availability of complementary components and availability of executable pieces of software</h2>
<p>In their editorial, Tsafnat et al.&nbsp;argue that i) the paradox of interoperability of having overlapping standards can be addressed by converging on just three standards; ii) practical and socio-technical considerations are as important as, if not more important than, technical superiority and therefore balancing of customizability and rigidity is of the essence; and iii) open standards, backed by engaged communities, hold an advantage over proprietary ones. While we concur with these points, we argue that these are necessary, but not sufficient conditions for convergence of health data standards. Existing research on digital platforms underlines the importance of the platform’s openness, not only in terms of open standards, but also in terms of availability of executable pieces of software, extensibility of the code base and availability of complements to the core technical platform (in this case the health data standard is a critical, defining component of the core technical platform) <span class="citation" data-cites="dereuver2018digital">[<a href="#ref-dereuver2018digital" role="doc-biblioref">2</a>]</span>. Openness in this context pertains to the software modules that constitute the digital platform. Realizing openness can be achieved through open sourcing the core components of the platform or defining standardized interfaces through which components can interact <span class="citation" data-cites="dereuver2022openness">[<a href="#ref-dereuver2022openness" role="doc-biblioref">6</a>]</span>. Only when the majority of these aspects of digital platforms are met can we reasonably expect that the digital platform will indeed flourish and be long lived.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Textbox 1: Conceptual background of the digital platform.
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>Digital platforms</strong> are software-based online infrastructures that facilitate interactions and transactions between users. In the context of this paper, digital platforms serve as an interface used to interact with data systems. <strong>Data systems</strong> describe a set of technologies, tools and processes that extract, manage and deliver data. Where the <strong>data system</strong> describes the functional implementation, the <strong>data architecture</strong> specifies the design framework, outlining how the data flows in its collection, storage, processing and governance. Its key components are <strong>data sources</strong> (original ‘raw’ data that is collected before any processing), data repositories like <strong>databases</strong>, <strong>data warehouses</strong> or <strong>lakes</strong> and data processing engines and pipelines that transform raw data into a usable format for analysis.</p>
<p>All architectures include a <strong>core technical platform</strong> (the foundational infrastructure) that can be extended to facilitate the necessary digital services. <strong>Data architectures</strong> contain different levels of specifications for the technical components entailed in the system. These levels include a systems <strong>code base</strong> (machine-readable text describing how to extract and process certain data), <strong>software tools</strong> (programs and applications enabling digital operations) and <strong>stacks</strong> (layers of software systems working together).</p>
</div>
</div>
<p>If open digital platforms are what we want, the question is how to achieve that. In what they frame as ‘the paradox of open’, Keller and Tarkowski argue that open platforms and their associated ecosystems can only flourish if two types of conditions are met <span class="citation" data-cites="keller2021paradox">[<a href="#ref-keller2021paradox" role="doc-biblioref">3</a>]</span>. The first condition states that many people need to contribute to the creation of a common resource. “This is the story of Wikipedia, OpenStreetMap, Blender.org, and the countless free software projects that provide much of the internet’s infrastructure.” <span class="citation" data-cites="keller2021paradox">[<a href="#ref-keller2021paradox" role="doc-biblioref">3</a>]</span> Indeed, Tsafnat et al.&nbsp;have explicitly taken into account that “an engaged and vibrant community is a major advantage for the longevity of the data standards it uses,” which has informed their proposal to converge towards OMOP, FHIR and openEHR over other existing health data standards. However, the importance of OSS is somewhat overlooked. This point is only mentioned in passing when Tsafnat et al.&nbsp;reference work done by Reynolds and Wyatt who already argued in 2011 “… for the superiority of open-source licensing to promote safer, more effective healthcare information systems. We claim that open-source licensing in health care information systems is essential to rational procurement strategy” <span class="citation" data-cites="reynolds2011open">[<a href="#ref-reynolds2011open" role="doc-biblioref">7</a>]</span>. Hence, we extend the line of reasoning of Tsafnat et al.&nbsp;by emphasizing that the availability of executable OSS components, which inherently make it easier to extend the code base of the health data standard and thereby driving greater availability of complementary components, is an important criterion which needs to be explicitly taken into account when choosing which standard to adopt.</p>
<p>The second condition put forward by Keller and Tarkowski is that open ecosystems have proven fruitful when “opening up” is the result of external incentives or requirements, rather than voluntary actions. Examples of such external incentives are “… publicly funded knowledge production like Open Access academic publications, cultural heritage collections in the Public Domain, Open Educational Resources, and Open Government data.” <span class="citation" data-cites="keller2021paradox">[<a href="#ref-keller2021paradox" role="doc-biblioref">3</a>]</span> Another canonical example is the birth of the GSM standard, which was mandated by European legislation <span class="citation" data-cites="wikipedia-gsm">[<a href="#ref-wikipedia-gsm" role="doc-biblioref">8</a>]</span>. Reflecting on this condition in the context of open health data ecosystems, we observe a salient difference between FHIR versus openEHR and OMOP, namely that the former is the only one that has been mandated - or at least strongly recommended - in some jurisdictions. Survey results on the state of FHIR show that the FHIR standard has been mandated or advised in 20 countries <span class="citation" data-cites="hl72024state">[<a href="#ref-hl72024state" role="doc-biblioref">9</a>]</span>. Notably, the European Electronic Health Record Exchange Format (EHRxF), introduced by the European Commission in 2019 with the aim to ensure secure, interoperable, cross-border access to electronic health data across the EU, decided in 2022 to adopt HL7 FHIR as the exchange format for future priority data categories <span class="citation" data-cites="carmo2024d22">[<a href="#ref-carmo2024d22" role="doc-biblioref">10</a>]</span>. In the US, the Office of the National Coordinator for Health Information Technology and the Centers for Medicare and Medicaid Services have introduced a steady stream of new regulations, criteria, and deadlines in Health IT that has resulted in significant adoption of FHIR <span class="citation" data-cites="firely2023fhir">[<a href="#ref-firely2023fhir" role="doc-biblioref">11</a>]</span>. In India, the open Health Claims Exchange protocol specification - which is based on FHIR - has been mandated by the Indian government as the standard for e-claims handling <span class="citation" data-cites="india2020national 2023hcx">[<a href="#ref-india2020national" role="doc-biblioref">12</a>,<a href="#ref-2023hcx" role="doc-biblioref">13</a>]</span>. The African Union recommends all new implementations and digital health system improvements use FHIR as the primary mechanism for data exchange <span class="citation" data-cites="tilahun2023african">[<a href="#ref-tilahun2023african" role="doc-biblioref">14</a>]</span>, but doesn’t say anything about the use of, for example, openEHR for clinical point-of-service systems.</p>
<p>Our third critical reflection on choosing health data standards pertains to the notion of the hourglass model <span class="citation" data-cites="estrin2010health beck2019hourglass">[<a href="#ref-estrin2010health" role="doc-biblioref">4</a>,<a href="#ref-beck2019hourglass" role="doc-biblioref">5</a>]</span> and the concept of open architectures <span class="citation" data-cites="mehl2023fullstac">[<a href="#ref-mehl2023fullstac" role="doc-biblioref">15</a>]</span>. The hourglass model is “… an approach to design that seeks to support a great diversity of applications (at the top of the hourglass) and allow implementation using a great diversity of supporting services (at the bottom).” <span class="citation" data-cites="beck2019hourglass">[<a href="#ref-beck2019hourglass" role="doc-biblioref">5</a>]</span> The center of the hourglass - the waist or also called the spanning layer in the information systems parlance - is defined by a set of minimal standards which mediates all interactions between the higher and lower layers. In the case of the Internet, the spanning layer is defined by the TCP/IP protocol, which is supported by a variety of underlying connectivity services (many different physical networks) on top of which many different applications can be built (email, videoconferencing etc.). We argue that FHIR has an added benefit over openEHR and OMOP because it can act as the spanning layer within an open health data platform. Because FHIR is inherently designed to function as a data exchange standard, it can function as a mediator between different components of the health data platform. The modularity of the various components that are part of the FHIR ecosystem allow it to be used effectively to implement subsystems.</p>
<p>We argue that i) the external incentives that have mandated FHIR in certain jurisdictions, and ii) the inherent modularity of the FHIR standard have resulted in a large boost in both commercial and OSS development activities in the FHIR ecosystem. Illustrative of this is the speed with which the Bulk FHIR API has been defined and implemented in almost all major implementations <span class="citation" data-cites="mandl2020push jones2021landscape">[<a href="#ref-mandl2020push" role="doc-biblioref">16</a>,<a href="#ref-jones2021landscape" role="doc-biblioref">17</a>]</span>, and the SQL-on-FHIR specification to make large-scale analysis of FHIR data both accessible to a larger audience as well as portable between systems <span class="citation" data-cites="sql-on-fhir-v2">[<a href="#ref-sql-on-fhir-v2" role="doc-biblioref">18</a>]</span>.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Textbox 2: Conceptual background of data processing pipelines for analytics.
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>Data pipelines</strong> define a sequence or workflow of processes for data. <strong>Data processing engines</strong> are tools that process, transform and analyze large-scale data and as such provide the foundational infrastructure to implement data pipelines. <strong>Computing workloads</strong> are specific tasks executed across data systems, like data processing and analytics.</p>
<p><strong>Data transformation</strong> entails all the processing pipelines that convert data into usable insights. <strong>Mappings</strong> are specific data transformations that aim to align data from different sources with a unified structure. <strong>Granular mappings</strong> transform data at the most detailed level, translating data elements across different schemas. <strong>Queries</strong> are built on top of transformed data, and retrieve data for insights generation, sometimes requiring further data processing.</p>
</div>
</div>
<p>The external incentives have also led to more people voluntarily contributing to FHIR-related OSS projects, which has resulted in a wide offering of FHIR components across major technology stacks (Java, Python, .NET), thereby strengthening the first condition for establishing openness. By comparison, OMOP and openEHR have profited less from external incentives to spur the adoption and thereby growing the ecosystem beyond a certain critical mass. To illustrate this, a quick-scan of the available OSS components listed on the website of the three governing bodies HL7 <span class="citation" data-cites="fhir-implementations">[<a href="#ref-fhir-implementations" role="doc-biblioref">19</a>]</span>, OHDSI <span class="citation" data-cites="ohdsi-implementations">[<a href="#ref-ohdsi-implementations" role="doc-biblioref">20</a>]</span> and openEHR <span class="citation" data-cites="openehr-implementations">[<a href="#ref-openehr-implementations" role="doc-biblioref">21</a>]</span>, indicates that the ecosystem of FHIR and OMOP have a significantly larger offering of extensible and complementary OSS components than openEHR, although for the latter notable a mature OSS implementation is available with EHRbase <span class="citation" data-cites="ehrbase">[<a href="#ref-ehrbase" role="doc-biblioref">22</a>]</span>. Taking GitHub as an proxy of worldwide development activities, <a href="#tbl-github" class="quarto-xref">Table&nbsp;1</a> shows the number of contributors and repositories for three different search terms. Note that these numbers should be taken as rough indicators. Given that the FHIR standard has more application areas, one would expect more GitHub projects than for openEHR.</p>
<div id="tbl-github" class="quarto-layout-panel anchored">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-github-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;1: Number of contributors and number of repositories on GitHub for the three healthcare data standards as per 28-01-2025.
</figcaption>
<div aria-describedby="tbl-github-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="tbl-github" style="flex-basis: 50.0%;justify-content: center;">
<div id="tbl-last-three-months" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-subfloat-tbl figure">
<figcaption class="quarto-float-caption-top quarto-subfloat-caption quarto-subfloat-tbl" id="tbl-last-three-months-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) Last three months
</figcaption>
<div aria-describedby="tbl-last-three-months-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<table class="caption-top table">
<thead>
<tr class="header">
<th style="text-align: left;">search term</th>
<th style="text-align: center;"># contrib.</th>
<th style="text-align: center;"># repos.</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">“openEHR”</td>
<td style="text-align: center;">82</td>
<td style="text-align: center;">49</td>
</tr>
<tr class="even">
<td style="text-align: left;">“OMOP” or “OHDSI”</td>
<td style="text-align: center;">446</td>
<td style="text-align: center;">221</td>
</tr>
<tr class="odd">
<td style="text-align: left;">“FHIR</td>
<td style="text-align: center;">1,648</td>
<td style="text-align: center;">756</td>
</tr>
</tbody>
</table>
</div>
</figure>
</div>
</div>
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="tbl-github" style="flex-basis: 50.0%;justify-content: center;">
<div id="tbl-second" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-subfloat-tbl figure">
<figcaption class="quarto-float-caption-top quarto-subfloat-caption quarto-subfloat-tbl" id="tbl-second-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) All time
</figcaption>
<div aria-describedby="tbl-second-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<table class="caption-top table">
<thead>
<tr class="header">
<th style="text-align: center;"># contrib.</th>
<th style="text-align: center;"># repos.</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">429</td>
<td style="text-align: center;">450</td>
</tr>
<tr class="even">
<td style="text-align: center;">1,019</td>
<td style="text-align: center;">113</td>
</tr>
<tr class="odd">
<td style="text-align: center;">8,497</td>
<td style="text-align: center;">8,617</td>
</tr>
</tbody>
</table>
</div>
</figure>
</div>
</div>
</div>
</div>
</figure>
</div>
<p>In summary, we stress that beyond evaluating the intrinsic structure of an open standard and the community that supports the standard, we need to consider the wider ecosystem of OSS implementations and availability of complementary components. From this wider perspective of the whole ecosystem surrounding the three standards, FHIR stands out as having the most diverse and rich ecosystem because it has been mandated in certain jurisdictions and because its technical foundations are inherently more broad and modular. This is relevant when comparing these standards in real-world implementations. We now turn to two specific use cases where these considerations are at play.</p>
</section>
<section id="standardization-of-health-data-for-federated-learning" class="level2">
<h2 class="anchored" data-anchor-id="standardization-of-health-data-for-federated-learning">Standardization of health data for federated learning</h2>
<p>The current fragmentation in health data is one of the major barriers towards leveraging the potential medical data for machine learning (ML). Without access to sufficient data, ML will be limited in its application to health improvement efforts and, ultimately, from making the transition from research to clinical practice. High quality health data, obtained from a research setting or a real-world clinical practice setting, is hard to obtain, because health data is tightly regulated.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Textbox 3: Conceptual background of distributed data systems.
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>Data systems</strong> often have a centralized architecture, where data is collected in a single repository or location. However, data systems can also <strong>distribute</strong> the storage and processing of data across different nodes or locations such as servers and edge devices.</p>
<p><strong>Servers</strong> act as the central processing units in data architecture, supporting computing workloads in data extraction, storage and transformation of data. <strong>Edge devices</strong> mainly provide support to the data extraction and preprocessing, generally located near the source of the data.</p>
<p><strong>Federated learning</strong> is an approach where machine learning models are trained across a distributed data system. Data transformations and analyses are performed on locally held data across multiple nodes, typically using edge devices or local servers. In this setup, the server that hosts the machine learning model does not need direct access to the source data. Instead, it aggregates the outputs of the local nodes (the updated model parameters) to train a global model. This method ensures that sensitive data remains local, preserving privacy while still enabling collaborative model training across distributed systems.</p>
</div>
</div>
<p>Federated learning (FL) is a learning paradigm that aims to address these issues of data governance and privacy by training algorithms collaboratively without moving (copying) the data itself <span class="citation" data-cites="rieke2020future teo2024federated">[<a href="#ref-rieke2020future" role="doc-biblioref">23</a>,<a href="#ref-teo2024federated" role="doc-biblioref">24</a>]</span>. Based on ongoing work with the PLUGIN healthcare consortium <span class="citation" data-cites="plugin">[<a href="#ref-plugin" role="doc-biblioref">25</a>]</span>, we have detailed an architecture for FL for secondary use of health data for hospitals in the Netherlands. The starting point for this implementation is the National Health Data Infrastructure agreements for research, policy and innovation for the Dutch healthcare sector, which have been adopted at the beginning of 2024 <span class="citation" data-cites="healthri2024agreements">[<a href="#ref-healthri2024agreements" role="doc-biblioref">26</a>]</span>. <a href="#fig-healthri-architecture" class="quarto-xref">Figure&nbsp;1</a> shows a high level reference architecture of the infrastructure to be, comprising three areas (multiple use, applications and generic features) and a total of 26 functional components <span class="citation" data-cites="healthri2024agreements">[<a href="#ref-healthri2024agreements" role="doc-biblioref">26</a>]</span>. One of the prerequisites of this architecture is that organizations that participate in a federation of ‘data stations’ use the same common data model to make the data Findable, Accessible, Interoperable and Resusable (FAIR). These FAIR data stations comprise components 7, 8 and 9 in <a href="#fig-healthri-architecture" class="quarto-xref">Figure&nbsp;1</a>, i.e.&nbsp;the data, metadata and APIs, respectively, through which the the data station can be accessed and used.</p>
<div id="fig-healthri-architecture" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-fig" id="fig-healthri-architecture-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: Reference architecture for the Dutch health data infrastructure for research and innovation <span class="citation" data-cites="healthri2024agreements">[<a href="#ref-healthri2024agreements" role="doc-biblioref">26</a>]</span>
</figcaption>
<div aria-describedby="fig-healthri-architecture-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="health-ri-architecture.png" class="img-fluid figure-img">
</div>
</figure>
</div>
<p>Following the line of reasoning of Tsafnat et al., OMOP would be the go-to standard for storing the longitudinal data in each of the data stations, where data is transformed from the original source (component 6), stored using a common data model (component 7) and properly annotated with metadata (component 8). Indeed, by now there are quite a real-world implementations of federated learning networks based on the OHDSI-OMOP stack, including a global infrastructure with 22 centers for COVID19 prediction models <span class="citation" data-cites="khalid2021standardized">[<a href="#ref-khalid2021standardized" role="doc-biblioref">27</a>]</span>, FeederNet in South Korea with 57 participating hospitals <span class="citation" data-cites="lee2022feedernet">[<a href="#ref-lee2022feedernet" role="doc-biblioref">28</a>]</span>, Dutch multi-cohort dementia research with 9 centers <span class="citation" data-cites="mateus2024data">[<a href="#ref-mateus2024data" role="doc-biblioref">29</a>]</span>, the European severe heterogeneous asthma research collaboration <span class="citation" data-cites="kroes2022blueprint">[<a href="#ref-kroes2022blueprint" role="doc-biblioref">30</a>]</span> and the recently initiated Belgian Federated Health Innovation Network (FHIN) <span class="citation" data-cites="deltomme2024federated">[<a href="#ref-deltomme2024federated" role="doc-biblioref">31</a>]</span>.</p>
<p>For the PLUGIN project, however, we choose to adopt FHIR as a data model because it is more compatible with the data model of the clinical administration systems. As PLUGIN focuses on secondary use of routine health data, we feel it is more suitable than OMOP, the latter being more suitable for clinical research data. OpenEHR might have been an option, too, if more implementations and complementary components had been available. Another reason for choosing FHIR is its practicality and extensibility to be used in a Python-based data science stack, provenance of RESTful APIs out-of-the-box to facilitate easy integration with the container-based vantage6 FL framework, and the support of many healthcare terminologies and flexibility through its profiling mechanism <span class="citation" data-cites="moncada-torres2021vantage6 choudhury2020personal smits2022improved">[<a href="#ref-moncada-torres2021vantage6" role="doc-biblioref">32</a>–<a href="#ref-smits2022improved" role="doc-biblioref">34</a>]</span>. Increasingly, other projects have reported the use of FHIR for persistent, longitudinal storage for FL. A scoping review on the use of FHIR for clinical research shows that it is increasingly being used for data preparation, cohort selection and secondary data sharing <span class="citation" data-cites="duda2022hl7">[<a href="#ref-duda2022hl7" role="doc-biblioref">35</a>]</span>. The CODA platform, which aims to implement a FL infrastructure in Canada similar to the PLUGIN project, compared OMOP and FHIR and chose the latter as it has been found to support more granular mappings required for analytics <span class="citation" data-cites="mullie2023coda">[<a href="#ref-mullie2023coda" role="doc-biblioref">36</a>]</span>. The fair4health project used FHIR as part of a FAIRification workflow to simplify the process of data extraction and preparation for clinical study analyses <span class="citation" data-cites="sinaci2024privacypreserving">[<a href="#ref-sinaci2024privacypreserving" role="doc-biblioref">37</a>]</span>.</p>
<p>Given that OMOP can, conceptually, be viewed as a strict subset of FHIR, hybrid solutions using a combination of OMOP and FHIR have also been reported, such as the German KETOS platform <span class="citation" data-cites="gruendner2019ketos">[<a href="#ref-gruendner2019ketos" role="doc-biblioref">38</a>]</span>, and the preliminary findings from the European GenoMed4All project which aims to connect clinical and -omics data <span class="citation" data-cites="cremonesi2023need">[<a href="#ref-cremonesi2023need" role="doc-biblioref">39</a>]</span>. A collaboration of 10 university hospitals in Germany have shown that standardized ETL-processing from FHIR into OMOP can achieve 99% conformance <span class="citation" data-cites="peng2023etlprocess">[<a href="#ref-peng2023etlprocess" role="doc-biblioref">40</a>]</span>, which confirms the feasibility of the solution pattern where FHIR acts as an intermediate sharing standard through which data from (legacy) systems are extracted and made available for reuse in a common data model. One could argue that the distinction between FHIR and OMOP becomes less relevant if data can be effectively stored in either standard. We are hopeful that initiatives like OMOP-on-FHIR indeed will foster convergence rather than collision between these two standards <span class="citation" data-cites="omoponfhir">[<a href="#ref-omoponfhir" role="doc-biblioref">41</a>]</span>.</p>
<p>In the case of PLUGIN, another important consideration for choosing FHIR over OMOP is, that from a data architecture perspective, the mechanism of FHIR Profiles can be tied to principle of late binding commonly applied in data lake/warehouse architectures (<a href="#fig-late-binding" class="quarto-xref">Figure&nbsp;2</a>): allow ingest of widely different sources, and gradually add more constraints and validations as you move closer to a specific use case. If machine learning is the primary objective for secondary use, one wants to be able to cast a wider net of relevant data, rather than being too restrictive when ingesting the data at the start of the processing pipeline. Late binding in data warehousing is a design philosophy where data transformation and schema enforcement are deferred as late as possible in the data processing pipeline, sometimes even until query time. This approach contrasts with early binding, where data is transformed and structured as it is ingested into the data warehouse. The advantage of this design is that it allows for greater flexibility and allows us to leverage new standards and technologies using the lakehouse architecture and the composable data stack for the implmentation of the data stations (see info box 4). During the initial ingestion of the data, we only require the data to conform to the minimal syntactic standard defined by the base FHIR version (R4 in the diagram). As the data is processed, more strict checks and constraints are applied, whereby ultimately different profiles can co-exists next to one another (the two most inner rectangles), within a larger rectangle with fewer restrictions. Note that if any of the profiles includes a FHIR extension, such as adding a field to include a birth name, the profiles are no longer strictly concentric. Hence extra care needs to be taken when dealing with extensions when applying the principle of late binding.</p>
<div id="fig-late-binding" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-fig" id="fig-late-binding-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2: Principle of late binding with FHIR profiling mechanism, illustrated with FHIR Profiles that are currently in use in the Netherlands.
</figcaption>
<div aria-describedby="fig-late-binding-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="late-binding.png" class="img-fluid figure-img">
</div>
</figure>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Textbox 4: Lakehouse architecture and the composable data stack
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>Data lakehouses</strong> typically have a zonal architecture that follows the Extract-Load-Transform (ELT) parttern where data is ingested from the source systems in bulk (E), delivered to storage with aligned schemas (L) and transformed into a format ready for analysis and re-use (T) <span class="citation" data-cites="hai2023data harby2022data harby2024data">[<a href="#ref-hai2023data" role="doc-biblioref">42</a>–<a href="#ref-harby2024data" role="doc-biblioref">44</a>]</span>. The discerning characteristic of the lakehouse architecture is its foundation on low-cost and directly-accessible storage that also provides traditional database management and performance features such as ACID transactions, data versioning, auditing, indexing, caching, and query optimization <span class="citation" data-cites="armbrust2021lakehouse">[<a href="#ref-armbrust2021lakehouse" role="doc-biblioref">45</a>]</span>.</p>
<p>The <strong>composable data stack</strong> <span class="citation" data-cites="pedreira2023composable">[<a href="#ref-pedreira2023composable" role="doc-biblioref">46</a>]</span> is a new set of technologies and open standards for fast processing of data using columnar data formats, including Apache Arrow as the standard columnar in-memory format with RPC-based data movement <span class="citation" data-cites="apache-arrow">[<a href="#ref-apache-arrow" role="doc-biblioref">47</a>]</span>; Apache Parquet as the standard columnar on-disk format <span class="citation" data-cites="apache-parquet">[<a href="#ref-apache-parquet" role="doc-biblioref">48</a>]</span>; and Apache Iceberg as the open table format <span class="citation" data-cites="jain2023analyzing apache-iceberg">[<a href="#ref-jain2023analyzing" role="doc-biblioref">49</a>,<a href="#ref-apache-iceberg" role="doc-biblioref">50</a>]</span>. This design also enables the use of new embedded, in-memory data processing engines. In turn, this opens up possibilities to bring computing workloads to edge devices, such as running DuckDB in the browser on top of WebAssembly <span class="citation" data-cites="duckdb">[<a href="#ref-duckdb" role="doc-biblioref">51</a>]</span>.</p>
<p>Using these technologies, full separation of storage and compute can be achieved which allows for cost-effective implementation of data stations.</p>
</div>
</div>
<p>One of the key challenges in using FHIR in this way pertains to the need for upgrading the whole ELT pipeline when upgrading to a new primary FHIR version, for example R6. The potential technical debt of version upgrades in the future is not specific to FHIR, but being a younger standard changes are more frequent compared to OMOP and openEHR. However, we expect that the development time required to upgrade FHIR versions is significantly less than the initial migration to FHIR.</p>
<p>The considerations above also show the conceptual difference of FHIR as a health data exchange standard versus openEHR as a persistent storage of routine healthcare data and OMOP as a persistent storage of health research data. For health data exchange and federated learning, the recipient of the data determines, to a large extent, what subset of data in the source needs to be made available – i.e.&nbsp;the target data model is known late and this favors late binding. In the case of routine collection of data, the holder of the source data determines what data needs to be stored – and typically everything – which favors early binding.</p>
</section>
<section id="health-data-standards-in-lmics" class="level2">
<h2 class="anchored" data-anchor-id="health-data-standards-in-lmics">Health data standards in LMICs</h2>
<p>It is a widely held belief that digital technologies have an important role to play in strengthening health systems in LMICs. Yet, also here the current fragmentation of health data stands in the way of scaling up digital health programs beyond project-centric, vertical solutions into sustainable health information exchanges <span class="citation" data-cites="karamagi2022ehealth">[<a href="#ref-karamagi2022ehealth" role="doc-biblioref">52</a>]</span>. In the context of global digital health developments, Mehl et al.&nbsp;have also called for convergence to open standards, similar to Tsafnat et al., but additionally stress the need for OSS (also our main argument in this paper), open content (representations of public health, health system or clinical knowledge to guide implementations) and open architectures (reusable enterprise architecture patterns for health systems) <span class="citation" data-cites="mehl2023fullstac">[<a href="#ref-mehl2023fullstac" role="doc-biblioref">15</a>]</span>. As for the open architecture, we see a convergence towards the OpenHIE framework <span class="citation" data-cites="openhie">[<a href="#ref-openhie" role="doc-biblioref">53</a>]</span>, which has been adopted by many African countries as the architectural blueprint for implementing nation-wide health information exchanges (HIE) <span class="citation" data-cites="mamuye2022health">[<a href="#ref-mamuye2022health" role="doc-biblioref">54</a>]</span>, including Nigeria <span class="citation" data-cites="dalhatu2023paper">[<a href="#ref-dalhatu2023paper" role="doc-biblioref">55</a>]</span>, Kenya <span class="citation" data-cites="thaiya2021adoption">[<a href="#ref-thaiya2021adoption" role="doc-biblioref">56</a>]</span> and Tanzania <span class="citation" data-cites="nsaghurwe2021one">[<a href="#ref-nsaghurwe2021one" role="doc-biblioref">57</a>]</span>. <a href="#fig-openhie" class="quarto-xref">Figure&nbsp;3</a> shows an overview of the OpenHIE architecture.</p>
<div id="fig-openhie" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-fig" id="fig-openhie-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3: OpenHIE architecture showing the Point of Service systems (black), the Interoperability Layer (green) and the Component Layer (blue).
</figcaption>
<div aria-describedby="fig-openhie-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="openhie.png" class="img-fluid figure-img">
</div>
</figure>
</div>
<p>While the OpenHIE specification is agnostic to which data standards should be used, in practice the digital health community in LMICs have converged towards FHIR as the primary standard for health information exchange, in line with the proposal by Tsafnat et al.&nbsp;To illustrate this point, consider the OpenHIM Platform architecture (<a href="#fig-openhim-platform" class="quarto-xref">Figure&nbsp;4</a>), which is currently the largest OSS implementation of the OpenHIE specification. In OpenHIM, clients (Point-of-Service systems) can initiate various workflows to submit or query patient data. The Shared Health Record (SHR) acts as the core transactional system for the health information exchange, which in this case is realized with the HAPI FHIR server, being one of the most widely used open-source FHIR server implementations <span class="citation" data-cites="hapi-fhir">[<a href="#ref-hapi-fhir" role="doc-biblioref">58</a>]</span>.</p>
<div id="fig-openhim-platform" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-fig" id="fig-openhim-platform-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4: OpenHIM Platform Architecture, illustrating the use of FHIR-based workflows between the components as specified in OpenHIE. CR: Client Registry. IOL: Interoperability Layer. MPI: Master Patient Index. SHR: Shared Health Record. Image source: <a href="https://jembi.gitbook.io/">https://jembi.gitbook.io/</a>.
</figcaption>
<div aria-describedby="fig-openhim-platform-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="openhim-platform.png" class="img-fluid figure-img">
</div>
</figure>
</div>
<p>Looking at the Point-of-Service systems, we see that as of today openEHR is rarely used as the standard for routine collection of clinical data in LMICs. The largest OSS electronic health record (EHR) implementations for low-resource settings are based on non-standardized data models, and it is unlikely this will change any time soon <span class="citation" data-cites="syzdykova2017opensource">[<a href="#ref-syzdykova2017opensource" role="doc-biblioref">59</a>]</span>. Instead, we see that FHIR-native software development frameworks such as OpenSRP <span class="citation" data-cites="mehl2020open">[<a href="#ref-mehl2020open" role="doc-biblioref">60</a>]</span> and the Open Health Stack <span class="citation" data-cites="open-health-stack">[<a href="#ref-open-health-stack" role="doc-biblioref">61</a>]</span> are being used more and more. In this approach, health professionals use Android apps to register and collect routine health data (<a href="#fig-opensrp" class="quarto-xref">Figure&nbsp;5</a>). As an example, OpenSRP has been deployed in 14 countries targeting various patient populations, amongst which is the implementation of the WHO antenatal and neonatal care guidelines for midwives in Lombok, Indonesia <span class="citation" data-cites="summitinstitutefordevelopment2023bunda kurniawan2019midwife">[<a href="#ref-summitinstitutefordevelopment2023bunda" role="doc-biblioref">62</a>,<a href="#ref-kurniawan2019midwife" role="doc-biblioref">63</a>]</span>. Beda EMR takes a similar approach and provides a FHIR native front-end that can be used in combination with any FHIR server as a backend <span class="citation" data-cites="beda-emr">[<a href="#ref-beda-emr" role="doc-biblioref">64</a>]</span>. Such a solution design is particularly useful for mid-size and smaller healthcare facilities, which are often resource constrained, lacking basic IT infrastructure to deploy a full-blown electronic medical record system. Hence the FHIR-based SHR functions as both, the administrative system-of-record and as the hub for information exchange at the same time.</p>
<div id="fig-opensrp" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-fig" id="fig-opensrp-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5: Overview of OpenSRP2 OSS framework for building clinical administration apps. HIS: health information systems. Image source: <a href="https://docs.opensrp.io/">https://docs.opensrp.io/</a>.
</figcaption>
<div aria-describedby="fig-opensrp-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="opensrp.png" class="img-fluid figure-img">
</div>
</figure>
</div>
<p>Finally, regarding longitudinal data analysis, we also see a convergence towards FHIR as the primary standard in LMICs. As in the case of federated learning, the choice for FHIR to implement data warehouse and analytic platforms is the preferred method due to the widespread availability of complementary OSS components. FHIR-specific technologies such as Bulk FHIR data access and SQL-on-FHIR mentioned earlier, allow the FHIR ecosystem to be used, complemented and integrated with generic OSS data warehousing components such as Clickhouse <span class="citation" data-cites="clickhouse">[<a href="#ref-clickhouse" role="doc-biblioref">65</a>]</span> and dbt <span class="citation" data-cites="dbt">[<a href="#ref-dbt" role="doc-biblioref">66</a>]</span>. Recently, more studies have pointed to the potential that FHIR brings when it is used in conjunction with machine learning and AI <span class="citation" data-cites="balch2023machine">[<a href="#ref-balch2023machine" role="doc-biblioref">67</a>]</span>. FHIR-based shared health records can act as systems of records for countries, thereby enabling reuse by health researchers, foundations, etc. to create public value with this data.</p>
<p>All in all, we see that in the context of LMICs, the standardization of the three domains put forward by Tsafnat merge into one. The SHR, as the key component within the OpenHIE specification, serves as the back-end of the system-of-record and provides a transactional, persistent storage engine for information exchange. Downstream longitudinal data stores continue to use FHIR as the common data model for analytical purposes. One could argue that it is in fact advantageous to converge to just one standard, thereby reducing complexity and cost of the total system. Such a perspective ties in with the notion of the hourglass model and open architectures: because FHIR is inherently designed to make optimal use of internet standards, such as the json file format and REST APIs, it is very modular and developer friendly. The many components that make up the FHIR allows the standard to be used effectively to implement subsystems, such as a facility registry or a health worker registry. By comparison, OMOP and openEHR are designed with a smaller scope with fewer application areas and are thereby less suitable as a standard to implement the subsystems defined in the OpenHIE specification.</p>
</section>
<section id="discussion-conclusion-and-future-research" class="level2">
<h2 class="anchored" data-anchor-id="discussion-conclusion-and-future-research">Discussion, conclusion and future research</h2>
<p>We agree with Tsafnat et al.&nbsp;that there is a dire need to converge to open data standards in healthcare, and support their proposal to focus on openEHR, FHIR and OMOP in healthcare informatics going forward. However, open standards are a necessary but not sufficient condition for the convergence of health data standardization. The availability of OSS implementations and complementary technologies is important when choosing which open standard to use. We find that the proposed trichotomy is too restrictive and therefore of limited use in guiding design choices to be made in real-world scenarios. Instead, we think that the full-STAC approach described by Mehl et al.&nbsp;is more comprehensive <span class="citation" data-cites="mehl2023fullstac">[<a href="#ref-mehl2023fullstac" role="doc-biblioref">15</a>]</span>. Furthermore, we argue that FHIR has the potential of acting as the spanning layer for health data interoperability, thereby enabling much wider standardization and adoption within the health data ecosystem at large. This is illustrated by the two cases considered in this paper, where FHIR is used beyond its original scope as a health data exchange standard.</p>
<p>In the case of FL, FHIR can be used interchangeably with OMOP for longitudinal analysis. Also, due to its inherently modular design, FHIR can be used in conjunction with the principle of late binding, as opposed to early binding for OMOP and openEHR, which is a relevant design criterion for implementing federated data platforms for secondary use. In the case of LMICs, we see that FHIR is emerging as the standard for all three domains of routine health data collection at the clinical point-of-service, data exchange and longitudinal analysis. We believe this is driven by the resource constrained setting in LMICs, the modularity of FHIR and the lower complexity and shallower learning curve of FHIR compared to openEHR. We expect that FHIR will play a major role in driving health data convergence in LMICs, because the availability of OSS implementations and complementary components are important enablers in these resource-constrained environments. We strongly support ongoing developments to increase the availability of OSS implementations as digital public goods <span class="citation" data-cites="digitalpublicgoods">[<a href="#ref-digitalpublicgoods" role="doc-biblioref">68</a>]</span> and integration projects such as Instant OpenHIE <span class="citation" data-cites="instant-openhie-v2">[<a href="#ref-instant-openhie-v2" role="doc-biblioref">69</a>]</span>, which will improve health data interoperability in LMICs.</p>
<p>Although openEHR has not been chosen as the standard for the two use cases presented here, we want to stress that it is not our intention to argue for or against any of the three standards <em>a priori</em> nor do we intend to dismiss openEHR outright. Instead, our aim is to illustrate the kind of design choices and trade-offs that need to be made, particularly those related to the availability and complementarity of OSS components. Significant developments and uptake of openEHR as a clinical data respository have been reported, with currently 17 openEHR solution providers that have been implemented in thousands clinics and research organizations worldwide <span class="citation" data-cites="delussu2024survey">[<a href="#ref-delussu2024survey" role="doc-biblioref">70</a>]</span>. Additionally work is underway to integrate openEHR with FHIR for data exchange <span class="citation" data-cites="fhirconnectspec openfhir">[<a href="#ref-fhirconnectspec" role="doc-biblioref">71</a>,<a href="#ref-openfhir" role="doc-biblioref">72</a>]</span>. Some experts agree that openEHR is the only specification that provides a comprehensive solution for building a standardized EHR <span class="citation" data-cites="pedrera-jimenez2023cana">[<a href="#ref-pedrera-jimenez2023cana" role="doc-biblioref">73</a>]</span>. Furthermore, openEHR is not only being deployed as a clinical system-of-record, but also as persistent clinical data repository for implementing national health information exchanges in European Nordic countries <span class="citation" data-cites="pohjonen2022norway">[<a href="#ref-pohjonen2022norway" role="doc-biblioref">74</a>]</span> and Slovenia <span class="citation" data-cites="bajric2023building">[<a href="#ref-bajric2023building" role="doc-biblioref">75</a>]</span>, which is very similar to the solution design of the SHR within the OpenHIE architecture presented here. An ongoing program in the south of the Netherlands has demonstrated a decentralized data sharing ecosystem using separate openEHR data stores, where federated queries are supported by the openEHR Archetype Query Language <span class="citation" data-cites="rsozuid-limburg2025demonstratie">[<a href="#ref-rsozuid-limburg2025demonstratie" role="doc-biblioref">76</a>]</span>.</p>
<p>The two cases allow us to reflect and revisit the key arguments of this paper, namely, the importance of OSS implementations and availability of complementary components for wide-scale adoption of health data standards. There is an important and equally complex interplay between OSS development and standardization, where OSS implementation can occur before, after or in parallel to standardization efforts <span class="citation" data-cites="wright2014open blind2019relationship">[<a href="#ref-wright2014open" role="doc-biblioref">77</a>,<a href="#ref-blind2019relationship" role="doc-biblioref">78</a>]</span>. Various studies have provided increasing evidence that OSS is a key success factor in driving software-related standardization <span class="citation" data-cites="blind2019relationship">[<a href="#ref-blind2019relationship" role="doc-biblioref">78</a>]</span>, and by extension we think it is critical when aiming to achieve data standardization. The history of how the DICOM imaging standard came to be is a good example how OSS development was pivotal in achieving wide-scale adoption of this standard <span class="citation" data-cites="erickson2005role nagy2007open">[<a href="#ref-erickson2005role" role="doc-biblioref">79</a>,<a href="#ref-nagy2007open" role="doc-biblioref">80</a>]</span>.</p>
<p>In contrast, the phenomena of forking, fragmentation and splintering are known to hinder an industry to consolidate towards a set of open standards <span class="citation" data-cites="simcoe2019forking">[<a href="#ref-simcoe2019forking" role="doc-biblioref">81</a>]</span>. Given the specific characteristics of data as an artefact, fragmentation is arguably the most relevant of these phenomena. De Reuver et al.&nbsp;expect fragmentation to persist for some time in the evolution of data platforms and associated ecosystems <span class="citation" data-cites="dereuver2022openness">[<a href="#ref-dereuver2022openness" role="doc-biblioref">6</a>]</span>. The case of the Unix operating system is an interesting example where fragmentation hampered standardization, next to market dynamics and issues related to intellectual property rights <span class="citation" data-cites="simcoe2019forking">[<a href="#ref-simcoe2019forking" role="doc-biblioref">81</a>]</span>.</p>
<p>But even when OSS has successfully contributed to ‘tip’ the healthcare industry to a set of health data standards, issues remain regarding sustainability of the OSS ecosystem itself. The market dynamics and economics of OSS ecosystems differ considerably from industry to industry: sustainability of OSS in the context of, say, the cultural and scientific heritage sector will be different to the challenges of OSS projects that are used as mission-critical components of open digital infrastructures worldwide. In case of the latter, underfunding is a critical issue and initiatives such as the German Sovereign Tech Agency have been launched to alleviate this <span class="citation" data-cites="2025sovereidn">[<a href="#ref-2025sovereidn" role="doc-biblioref">82</a>]</span>. In the context of open health data standards, we believe that risks related to underfunding are lower and more manageable. Within the digital health community there is a range of commercial companies supporting the OSS projects and creating sustainable business from it using various business models like offering support contracts, split licensing and complementary closed source products <span class="citation" data-cites="chang2007open">[<a href="#ref-chang2007open" role="doc-biblioref">83</a>]</span>. Regarding the dynamics of forking and fragmentation mentioned earlier, we feel that code forking on balance has a net positive effect on long-term sustainability of OSS at the level of the software itself, the community and the ecosystem <span class="citation" data-cites="nyman2013code">[<a href="#ref-nyman2013code" role="doc-biblioref">84</a>]</span>.</p>
<p>Going forward, we suggest the following directions for future research. Given that health data standardization will continue to require mappings, we propose to explore the use of machine learning, and particularly large-language models, as a means to reduce the development effort required to create transformations between various health data formats. New machine learning methods can also be developed to assess and improve data quality across the various stages of the data processing pipelines. In terms of data integration, we expect that health data will increasingly be used in conjunction with data from social services and the welfare domain, which requires new techniques to integrate different data domains, for example using knowledge graphs and ontologies. Last, but certainly not least, future research should not only explore the technical but also the social implications of implementing OSS components for data standardization across the healthcare system, specifically in settings where governance or ethical considerations of data interoperability have not specifically been addressed at a regulatory level. In line with the embedding of open standards in the open-source ecosystem, we assert that the benefits of health data standardization will only be realized if they are coupled with collaborative, community-driven governance models. It remains essential to ensure that the development, adoption, and evolution of standards remain inclusive, transparent, and responsive to the diverse needs within the health system.</p>
</section>
<section id="authors-contributions" class="level2">
<h2 class="anchored" data-anchor-id="authors-contributions">Authors’ Contributions</h2>
<p>DK contributed to the concept and design of the manuscript and prepared the first draft. AD, MS and BJV contributed to the section on federated learning. FH and MB contributed to the section on LMICs. DK and FH revised the manuscript based on the feedback from the peer reviewers. All authors contributed to the final revision and approved the final manuscript.</p>
</section>
<section id="acknowledgements" class="level2">
<h2 class="anchored" data-anchor-id="acknowledgements">Acknowledgements</h2>
<p>The authors thankfully acknowledge Joost Holslag for his feedback and discussion on openEHR.</p>
</section>
<section id="conflicts-of-interests" class="level2">
<h2 class="anchored" data-anchor-id="conflicts-of-interests">Conflicts of interests</h2>
<p>DK received funding from PharmAccess as a contractor to conduct the work on LMICs reported here. MB/Ona is the core developer of the open-source OpenSRP 2 framework.</p>
</section>
<section id="abbreviations" class="level2">
<h2 class="anchored" data-anchor-id="abbreviations">Abbreviations</h2>
<p>ACID: Atomicity, Consistency, Isolation, Durability<br> API: Application Programming Interface<br> EHDS: European Health Data Space<br> EHR: Electronic Health Record<br> ELT: Extract, Load, Transform<br> FAIR: Findability, Accessibility, Interoperability, and Reusability<br> FHIR: Fast Healthcare Interoperability Resources<br> FL: Federated Learning<br> Full-STAC: Concept that advocates for open standards, open technology, open architecture and open content<br> GSM: Global System for Mobile Communications<br> HIE: Health Information Exchange<br> HL7: Health Level 7<br> LMIC: Low and middle-income countries<br> ML: Machine Learning<br> OHDSI: Observational Health Data Sciences and Informatics<br> OMOP: Observational Medical Outcomes Partnership<br> OSS: Open Source Software<br> SHR: Shared Health Record<br> REST: representational state transfer<br> TCP/IP: Transmission Control Protocol/Internet Protocol<br></p>
</section>
<section id="bibliography" class="level1 unnumbered">


</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body" role="list">
<div id="ref-tsafnat2024converge" class="csl-entry" role="listitem">
<div class="csl-left-margin">1. </div><div class="csl-right-inline">Tsafnat G, Dunscombe R, Gabriel D, Grieve G, Reich C. Converge or <span>Collide</span>? <span>Making Sense</span> of a <span>Plethora</span> of <span>Open Data Standards</span> in <span>Health Care</span>. <em>Journal of Medical Internet Research</em>. 2024;26(1):e55779. doi:<a href="https://doi.org/10.2196/55779">10.2196/55779</a></div>
</div>
<div id="ref-dereuver2018digital" class="csl-entry" role="listitem">
<div class="csl-left-margin">2. </div><div class="csl-right-inline">de Reuver M, Sørensen C, Basole RC. The <span>Digital Platform</span>: <span>A Research Agenda</span>. <em>Journal of Information Technology</em>. 2018;33(2):124-135. doi:<a href="https://doi.org/10.1057/s41265-016-0033-3">10.1057/s41265-016-0033-3</a></div>
</div>
<div id="ref-keller2021paradox" class="csl-entry" role="listitem">
<div class="csl-left-margin">3. </div><div class="csl-right-inline">Keller P, Tarkowski A. The <span>Paradox</span> of <span>Open</span>. <em>Open Future</em>. Published online March 5, 2021. Accessed March 25, 2024. <a href="https://openfuture.pubpub.org/pub/paradox-of-open/release/1">https://openfuture.pubpub.org/pub/paradox-of-open/release/1</a></div>
</div>
<div id="ref-estrin2010health" class="csl-entry" role="listitem">
<div class="csl-left-margin">4. </div><div class="csl-right-inline">Estrin D, Sim I. Health care delivery. <span class="nocase">Open mHealth</span> architecture: An engine for health care innovation. <em>Science</em>. 2010;330(6005):759-760. doi:<a href="https://doi.org/10.1126/science.1196187">10.1126/science.1196187</a></div>
</div>
<div id="ref-beck2019hourglass" class="csl-entry" role="listitem">
<div class="csl-left-margin">5. </div><div class="csl-right-inline">Beck M. On the hourglass model. <em>Communications of the ACM</em>. 2019;62(7):48-57. doi:<a href="https://doi.org/10.1145/3274770">10.1145/3274770</a></div>
</div>
<div id="ref-dereuver2022openness" class="csl-entry" role="listitem">
<div class="csl-left-margin">6. </div><div class="csl-right-inline">de Reuver M, Ofe H, Agahari W, Abbas AE, Zuiderwijk A. The openness of data platforms: A research agenda. In: <em>Proceedings of the 1st <span>International Workshop</span> on <span>Data Economy</span></em>. <span>DE</span> ’22. Association for Computing Machinery; 2022:34-41. doi:<a href="https://doi.org/10.1145/3565011.3569056">10.1145/3565011.3569056</a></div>
</div>
<div id="ref-reynolds2011open" class="csl-entry" role="listitem">
<div class="csl-left-margin">7. </div><div class="csl-right-inline">Reynolds CJ, Wyatt JC. Open <span>Source</span>, <span>Open Standards</span>, and <span>Health Care Information Systems</span>. <em>Journal of Medical Internet Research</em>. 2011;13(1):e1521. doi:<a href="https://doi.org/10.2196/jmir.1521">10.2196/jmir.1521</a></div>
</div>
<div id="ref-wikipedia-gsm" class="csl-entry" role="listitem">
<div class="csl-left-margin">8. </div><div class="csl-right-inline"><span>GSM</span>. In: <em>Wikipedia</em>.; 2024. Accessed September 20, 2024. <a href="https://en.wikipedia.org/w/index.php?title=GSM&amp;oldid=1245675274">https://en.wikipedia.org/w/index.php?title=GSM&amp;oldid=1245675274</a></div>
</div>
<div id="ref-hl72024state" class="csl-entry" role="listitem">
<div class="csl-left-margin">9. </div><div class="csl-right-inline"><em>The <span>State</span> of <span>FHIR</span> 2024 <span>Survey Results</span></em>. HL7; 2024. Accessed April 4, 2024. <a href="https://www.hl7.org/documentcenter/public/white-papers/2024%20StateofFHIRSurveyResults_final.pdf">https://www.hl7.org/documentcenter/public/white-papers/2024%20StateofFHIRSurveyResults_final.pdf</a></div>
</div>
<div id="ref-carmo2024d22" class="csl-entry" role="listitem">
<div class="csl-left-margin">10. </div><div class="csl-right-inline">Carmo A, Martins H. <em>D2.2 - <span>EHRxF</span> in a Nutshell-<span>WP2-ISCTE</span></em>.; 2024. <a href="https://ehr-exchange-format.eu/wp-content/uploads/2024/10/D2.2-v20240704-EHRxF-in-a-Nutshell-WP2-ISCTE.pdf">https://ehr-exchange-format.eu/wp-content/uploads/2024/10/D2.2-v20240704-EHRxF-in-a-Nutshell-WP2-ISCTE.pdf</a></div>
</div>
<div id="ref-firely2023fhir" class="csl-entry" role="listitem">
<div class="csl-left-margin">11. </div><div class="csl-right-inline">Firely. <em><span>FHIR</span> in <span>US</span> Healthcare Regulations</em>.; 2023. Accessed May 30, 2024. <a href="https://simplifier.net/organization/firely/news/153">https://simplifier.net/organization/firely/news/153</a></div>
</div>
<div id="ref-india2020national" class="csl-entry" role="listitem">
<div class="csl-left-margin">12. </div><div class="csl-right-inline"><em>National <span>Digital Health Mission</span></em>. India National Health Authority; 2020.</div>
</div>
<div id="ref-2023hcx" class="csl-entry" role="listitem">
<div class="csl-left-margin">13. </div><div class="csl-right-inline"><em><span>HCX Protocol</span> V0.9</em>.; 2023. Accessed September 18, 2024. <a href="http://hcxprotocol.io/">http://hcxprotocol.io/</a></div>
</div>
<div id="ref-tilahun2023african" class="csl-entry" role="listitem">
<div class="csl-left-margin">14. </div><div class="csl-right-inline">Tilahun B, Mamuye A, Yilma T, Shehata Y. <em>African <span>Union Health Information Exchange Guidelines</span> and <span>Standards</span></em>.; 2023.</div>
</div>
<div id="ref-mehl2023fullstac" class="csl-entry" role="listitem">
<div class="csl-left-margin">15. </div><div class="csl-right-inline">Mehl GL, Seneviratne MG, Berg ML, et al. A full-<span>STAC</span> remedy for global digital health transformation: Open standards, technologies, architectures and content. <em>Oxford Open Digital Health</em>. 2023;1:oqad018. doi:<a href="https://doi.org/10.1093/oodh/oqad018">10.1093/oodh/oqad018</a></div>
</div>
<div id="ref-mandl2020push" class="csl-entry" role="listitem">
<div class="csl-left-margin">16. </div><div class="csl-right-inline">Mandl KD, Gottlieb D, Mandel JC, et al. Push <span>Button Population Health</span>: <span>The SMART</span>/<span>HL7 FHIR Bulk Data Access Application Programming Interface</span>. <em>npj Digital Medicine</em>. 2020;3(1):1-9. doi:<a href="https://doi.org/10.1038/s41746-020-00358-4">10.1038/s41746-020-00358-4</a></div>
</div>
<div id="ref-jones2021landscape" class="csl-entry" role="listitem">
<div class="csl-left-margin">17. </div><div class="csl-right-inline">Jones J, Gottlieb D, Mandel JC, et al. A landscape survey of planned <span>SMART</span>/<span>HL7</span> bulk <span>FHIR</span> data access <span>API</span> implementations and tools. <em>Journal of the American Medical Informatics Association</em>. 2021;28(6):1284-1287. doi:<a href="https://doi.org/10.1093/jamia/ocab028">10.1093/jamia/ocab028</a></div>
</div>
<div id="ref-sql-on-fhir-v2" class="csl-entry" role="listitem">
<div class="csl-left-margin">18. </div><div class="csl-right-inline"><em><span>SQL</span> on <span>FHIR</span> V2.0.0-Pre</em>. Accessed September 20, 2024. <a href="https://build.fhir.org/ig/FHIR/sql-on-fhir-v2/">https://build.fhir.org/ig/FHIR/sql-on-fhir-v2/</a></div>
</div>
<div id="ref-fhir-implementations" class="csl-entry" role="listitem">
<div class="csl-left-margin">19. </div><div class="csl-right-inline"><span>FHIR Open Source Implementations</span>. September 20, 2024. Accessed September 20, 2024. <a href="https://confluence.hl7.org/display/FHIR/Open+Source+Implementations">https://confluence.hl7.org/display/FHIR/Open+Source+Implementations</a></div>
</div>
<div id="ref-ohdsi-implementations" class="csl-entry" role="listitem">
<div class="csl-left-margin">20. </div><div class="csl-right-inline">Software <span>Tools</span> – <span>OHDSI</span>. Accessed September 20, 2024. <a href="https://www.ohdsi.org/software-tools/">https://www.ohdsi.org/software-tools/</a></div>
</div>
<div id="ref-openehr-implementations" class="csl-entry" role="listitem">
<div class="csl-left-margin">21. </div><div class="csl-right-inline">Beale SH Thomas. <span class="nocase">openEHR Platform</span>. Accessed September 20, 2024. <a href="https://openehr.org/products_tools/platform/">https://openehr.org/products_tools/platform/</a></div>
</div>
<div id="ref-ehrbase" class="csl-entry" role="listitem">
<div class="csl-left-margin">22. </div><div class="csl-right-inline"><span>EHRbase</span> 2.0 website. Published online March 19, 2024. Accessed September 20, 2024. <a href="https://www.ehrbase.org/">https://www.ehrbase.org/</a></div>
</div>
<div id="ref-rieke2020future" class="csl-entry" role="listitem">
<div class="csl-left-margin">23. </div><div class="csl-right-inline">Rieke N, Hancox J, Li W, et al. The future of digital health with federated learning. <em>npj Digit Med</em>. 2020;3(1, 1):1-7. doi:<a href="https://doi.org/10.1038/s41746-020-00323-1">10.1038/s41746-020-00323-1</a></div>
</div>
<div id="ref-teo2024federated" class="csl-entry" role="listitem">
<div class="csl-left-margin">24. </div><div class="csl-right-inline">Teo ZL, Jin L, Liu N, et al. Federated machine learning in healthcare: <span>A</span> systematic review on clinical applications and technical architecture. <em>Cell Reports Medicine</em>. 2024;5(2):101419. doi:<a href="https://doi.org/10.1016/j.xcrm.2024.101419">10.1016/j.xcrm.2024.101419</a></div>
</div>
<div id="ref-plugin" class="csl-entry" role="listitem">
<div class="csl-left-margin">25. </div><div class="csl-right-inline"><span>PLUGIN</span> – <span>Platform</span> voor <span>Uitwisseling</span> en <span>Hergebruik</span> van <span>Klinische Data Nederland</span>. Accessed September 21, 2024. <a href="https://plugin.healthcare/">https://plugin.healthcare/</a></div>
</div>
<div id="ref-healthri2024agreements" class="csl-entry" role="listitem">
<div class="csl-left-margin">26. </div><div class="csl-right-inline">Health-RI. Agreements on the <span>National Health Data Infrastructure</span> for <span>Research</span>, <span>Policy</span> and <span>Innovation</span> - <span class="nocase">Health-RI Nationale Gezondheidsdata-infrastructuur</span> - <span>Confluence</span>. January 29, 2024. Accessed June 3, 2024. <a href="https://health-ri.atlassian.net/wiki/spaces/HNG/pages/249073646/Agreements+on+the+National+Health+Data+Infrastructure+for+Research+Policy+and+Innovation">https://health-ri.atlassian.net/wiki/spaces/HNG/pages/249073646/Agreements+on+the+National+Health+Data+Infrastructure+for+Research+Policy+and+Innovation</a></div>
</div>
<div id="ref-khalid2021standardized" class="csl-entry" role="listitem">
<div class="csl-left-margin">27. </div><div class="csl-right-inline">Khalid S, Yang C, Blacketer C, et al. A standardized analytics pipeline for reliable and rapid development and validation of prediction models using observational health data. <em>Computer Methods and Programs in Biomedicine</em>. 2021;211:106394. doi:<a href="https://doi.org/10.1016/j.cmpb.2021.106394">10.1016/j.cmpb.2021.106394</a></div>
</div>
<div id="ref-lee2022feedernet" class="csl-entry" role="listitem">
<div class="csl-left-margin">28. </div><div class="csl-right-inline">Lee S, Kim C, Chang J, Park RW. <span>FeederNet</span> (<span>Federated E-Health Big Data</span> for <span>Evidence Renovation Network</span>) platform in <span>Korea</span> – <span>OHDSI</span>. 2022. Accessed June 4, 2024. <a href="https://www.ohdsi.org/2022showcase-33/">https://www.ohdsi.org/2022showcase-33/</a></div>
</div>
<div id="ref-mateus2024data" class="csl-entry" role="listitem">
<div class="csl-left-margin">29. </div><div class="csl-right-inline">Mateus P, Moonen J, Beran M, et al. Data harmonization and federated learning for multi-cohort dementia research using the <span>OMOP</span> common data model: <span>A Netherlands</span> consortium of dementia cohorts case study. <em>Journal of Biomedical Informatics</em>. 2024;155:104661. doi:<a href="https://doi.org/10.1016/j.jbi.2024.104661">10.1016/j.jbi.2024.104661</a></div>
</div>
<div id="ref-kroes2022blueprint" class="csl-entry" role="listitem">
<div class="csl-left-margin">30. </div><div class="csl-right-inline">Kroes JA, Bansal AT, Berret E, et al. Blueprint for harmonising unstandardised disease registries to allow federated data analysis: Prepare for the future. <em>ERJ Open Research</em>. 2022;8(4). doi:<a href="https://doi.org/10.1183/23120541.00168-2022">10.1183/23120541.00168-2022</a></div>
</div>
<div id="ref-deltomme2024federated" class="csl-entry" role="listitem">
<div class="csl-left-margin">31. </div><div class="csl-right-inline">Deltomme C, Denturck K, De Jaeger P, et al. Federated <span>Health Innovation Network</span> (<span>FHIN</span>). Published online September 20, 2024. <a href="https://www.ohdsi-europe.org/images/symposium-2024/Posters/poster%20OHDSI%20FHIN%20Camille%20Deltomme%20-%20Camille%20Deltomme.pdf">https://www.ohdsi-europe.org/images/symposium-2024/Posters/poster%20OHDSI%20FHIN%20Camille%20Deltomme%20-%20Camille%20Deltomme.pdf</a></div>
</div>
<div id="ref-moncada-torres2021vantage6" class="csl-entry" role="listitem">
<div class="csl-left-margin">32. </div><div class="csl-right-inline">Moncada-Torres A, Martin F, Sieswerda M, Van Soest J, Geleijnse G. <span>VANTAGE6</span>: An open source <span class="nocase">priVAcy preserviNg federaTed leArninG infrastructurE</span> for <span class="nocase">Secure Insight eXchange</span>. <em>AMIA Annu Symp Proc</em>. 2021;2020:870-877. Accessed September 21, 2024. <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8075508/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8075508/</a></div>
</div>
<div id="ref-choudhury2020personal" class="csl-entry" role="listitem">
<div class="csl-left-margin">33. </div><div class="csl-right-inline">Choudhury A, van Soest J, Nayak S, Dekker A. Personal <span>Health Train</span> on <span>FHIR</span>: <span>A Privacy Preserving Federated Approach</span> for <span>Analyzing FAIR Data</span> in <span>Healthcare</span>. In: Bhattacharjee A, Borgohain SKr, Soni B, Verma G, Gao XZ, eds. <em>Machine <span>Learning</span>, <span>Image Processing</span>, <span>Network Security</span> and <span>Data Sciences</span></em>. Communications in <span>Computer</span> and <span>Information Science</span>. Springer; 2020:85-95. doi:<a href="https://doi.org/10.1007/978-981-15-6315-7_7">10.1007/978-981-15-6315-7_7</a></div>
</div>
<div id="ref-smits2022improved" class="csl-entry" role="listitem">
<div class="csl-left-margin">34. </div><div class="csl-right-inline">Smits D, Van Beusekom B, Martin F, Veen L, Geleijnse G, Moncada-Torres A. An <span>Improved Infrastructure</span> for <span>Privacy-Preserving Analysis</span> of <span>Patient Data</span>. In: Mantas J, Gallos P, Zoulias E, et al., eds. <em>Studies in <span>Health Technology</span> and <span>Informatics</span></em>. IOS Press; 2022. doi:<a href="https://doi.org/10.3233/SHTI220682">10.3233/SHTI220682</a></div>
</div>
<div id="ref-duda2022hl7" class="csl-entry" role="listitem">
<div class="csl-left-margin">35. </div><div class="csl-right-inline">Duda SN, Kennedy N, Conway D, et al. <span class="nocase">HL7 FHIR-based</span> tools and initiatives to support clinical research: A scoping review. <em>Journal of the American Medical Informatics Association</em>. 2022;29(9):1642-1653. doi:<a href="https://doi.org/10.1093/jamia/ocac105">10.1093/jamia/ocac105</a></div>
</div>
<div id="ref-mullie2023coda" class="csl-entry" role="listitem">
<div class="csl-left-margin">36. </div><div class="csl-right-inline">Mullie L, Afilalo J, Archambault P, et al. <span>CODA</span>: An open-source platform for federated analysis and machine learning on distributed healthcare data. <em>Journal of the American Medical Informatics Association</em>. Published online December 21, 2023:ocad235. doi:<a href="https://doi.org/10.1093/jamia/ocad235">10.1093/jamia/ocad235</a></div>
</div>
<div id="ref-sinaci2024privacypreserving" class="csl-entry" role="listitem">
<div class="csl-left-margin">37. </div><div class="csl-right-inline">Sinaci AA, Gencturk M, Alvarez-Romero C, et al. Privacy-preserving federated machine learning on <span>FAIR</span> health data: <span>A</span> real-world application. <em>Computational and Structural Biotechnology Journal</em>. 2024;24:136-145. doi:<a href="https://doi.org/10.1016/j.csbj.2024.02.014">10.1016/j.csbj.2024.02.014</a></div>
</div>
<div id="ref-gruendner2019ketos" class="csl-entry" role="listitem">
<div class="csl-left-margin">38. </div><div class="csl-right-inline">Gruendner J, Schwachhofer T, Sippl P, et al. <span>KETOS</span>: <span>Clinical</span> decision support and machine learning as a service – <span>A</span> training and deployment platform based on <span>Docker</span>, <span>OMOP-CDM</span>, and <span>FHIR Web Services</span>. <em>PLOS ONE</em>. 2019;14(10):e0223010. doi:<a href="https://doi.org/10.1371/journal.pone.0223010">10.1371/journal.pone.0223010</a></div>
</div>
<div id="ref-cremonesi2023need" class="csl-entry" role="listitem">
<div class="csl-left-margin">39. </div><div class="csl-right-inline">Cremonesi F, Planat V, Kalokyri V, et al. The need for multimodal health data modeling: <span>A</span> practical approach for a federated-learning healthcare platform. <em>Journal of Biomedical Informatics</em>. 2023;141:104338. doi:<a href="https://doi.org/10.1016/j.jbi.2023.104338">10.1016/j.jbi.2023.104338</a></div>
</div>
<div id="ref-peng2023etlprocess" class="csl-entry" role="listitem">
<div class="csl-left-margin">40. </div><div class="csl-right-inline">Peng Y, Henke E, Reinecke I, Zoch M, Sedlmayr M, Bathelt F. An <span class="nocase">ETL-process</span> design for data harmonization to participate in international research with <span>German</span> real-world data based on <span>FHIR</span> and <span>OMOP CDM</span>. <em>International Journal of Medical Informatics</em>. 2023;169:104925. doi:<a href="https://doi.org/10.1016/j.ijmedinf.2022.104925">10.1016/j.ijmedinf.2022.104925</a></div>
</div>
<div id="ref-omoponfhir" class="csl-entry" role="listitem">
<div class="csl-left-margin">41. </div><div class="csl-right-inline"><span>OMOPonFHIR</span>. Accessed September 20, 2024. <a href="https://omoponfhir.org/">https://omoponfhir.org/</a></div>
</div>
<div id="ref-hai2023data" class="csl-entry" role="listitem">
<div class="csl-left-margin">42. </div><div class="csl-right-inline">Hai R, Koutras C, Quix C, Jarke M. Data <span>Lakes</span>: <span>A Survey</span> of <span>Functions</span> and <span>Systems</span>. <em>IEEE Transactions on Knowledge and Data Engineering</em>. 2023;35(12):12571-12590. doi:<a href="https://doi.org/10.1109/TKDE.2023.3270101">10.1109/TKDE.2023.3270101</a></div>
</div>
<div id="ref-harby2022data" class="csl-entry" role="listitem">
<div class="csl-left-margin">43. </div><div class="csl-right-inline">Harby AA, Zulkernine F. From <span>Data Warehouse</span> to <span>Lakehouse</span>: <span>A Comparative Review</span>. In: <em>2022 <span>IEEE International Conference</span> on <span>Big Data</span> (<span>Big Data</span>)</em>. IEEE; 2022:389-395. doi:<a href="https://doi.org/10.1109/BigData55660.2022.10020719">10.1109/BigData55660.2022.10020719</a></div>
</div>
<div id="ref-harby2024data" class="csl-entry" role="listitem">
<div class="csl-left-margin">44. </div><div class="csl-right-inline">Harby AA, Zulkernine F. Data <span>Lakehouse</span>: <span>A Survey</span> and <span>Experimental Study</span>. doi:<a href="https://doi.org/10.2139/ssrn.4765588">10.2139/ssrn.4765588</a></div>
</div>
<div id="ref-armbrust2021lakehouse" class="csl-entry" role="listitem">
<div class="csl-left-margin">45. </div><div class="csl-right-inline">Armbrust M, Ghodsi A, Xin R, Zaharia M. Lakehouse: <span>A New Generation</span> of <span>Open Platforms</span> that <span>Unify Data Warehousing</span> and <span>Advanced Analytics</span>. In:; 2021:8.</div>
</div>
<div id="ref-pedreira2023composable" class="csl-entry" role="listitem">
<div class="csl-left-margin">46. </div><div class="csl-right-inline">Pedreira P, Erling O, Karanasos K, et al. The <span>Composable Data Management System Manifesto</span>. <em>Proc VLDB Endow</em>. 2023;16(10):2679-2685. doi:<a href="https://doi.org/10.14778/3603581.3603604">10.14778/3603581.3603604</a></div>
</div>
<div id="ref-apache-arrow" class="csl-entry" role="listitem">
<div class="csl-left-margin">47. </div><div class="csl-right-inline"><em>Apache <span>Arrow</span></em>.; 2024. Accessed September 20, 2024. <a href="https://arrow.apache.org/">https://arrow.apache.org/</a></div>
</div>
<div id="ref-apache-parquet" class="csl-entry" role="listitem">
<div class="csl-left-margin">48. </div><div class="csl-right-inline"><em>Apache <span>Parquet</span></em>.; 2024. Accessed September 20, 2024. <a href="https://parquet.apache.org/">https://parquet.apache.org/</a></div>
</div>
<div id="ref-jain2023analyzing" class="csl-entry" role="listitem">
<div class="csl-left-margin">49. </div><div class="csl-right-inline">Jain P, Kraft P, Power C, Das T, Stoica I, Zaharia M. Analyzing and <span>Comparing Lakehouse Storage Systems</span>. Published online 2023.</div>
</div>
<div id="ref-apache-iceberg" class="csl-entry" role="listitem">
<div class="csl-left-margin">50. </div><div class="csl-right-inline"><em>Apache <span>Iceberg</span></em>. Accessed September 20, 2024. <a href="https://iceberg.apache.org/">https://iceberg.apache.org/</a></div>
</div>
<div id="ref-duckdb" class="csl-entry" role="listitem">
<div class="csl-left-margin">51. </div><div class="csl-right-inline">User G. An in-process <span>SQL OLAP</span> database management system. Accessed October 10, 2024. <a href="https://duckdb.org/">https://duckdb.org/</a></div>
</div>
<div id="ref-karamagi2022ehealth" class="csl-entry" role="listitem">
<div class="csl-left-margin">52. </div><div class="csl-right-inline">Karamagi HC, Muneene D, Droti B, et al. <span class="nocase">eHealth</span> or e-<span>Chaos</span>: <span>The</span> use of <span>Digital Health Interventions</span> for <span>Health Systems Strengthening</span> in sub-<span>Saharan Africa</span> over the last 10 years: <span>A</span> scoping review. <em>J Glob Health</em>. 2022;12:04090. doi:<a href="https://doi.org/10.7189/jogh.12.04090">10.7189/jogh.12.04090</a></div>
</div>
<div id="ref-openhie" class="csl-entry" role="listitem">
<div class="csl-left-margin">53. </div><div class="csl-right-inline"><em><span>OpenHIE Framework</span> V5.2-En</em>.; 2024. Accessed August 27, 2024. <a href="https://ohie.org/">https://ohie.org/</a></div>
</div>
<div id="ref-mamuye2022health" class="csl-entry" role="listitem">
<div class="csl-left-margin">54. </div><div class="csl-right-inline">Mamuye AL, Yilma TM, Abdulwahab A, et al. Health information exchange policy and standards for digital health systems in africa: <span>A</span> systematic review. <em>PLOS Digital Health</em>. 2022;1(10):e0000118. doi:<a href="https://doi.org/10.1371/journal.pdig.0000118">10.1371/journal.pdig.0000118</a></div>
</div>
<div id="ref-dalhatu2023paper" class="csl-entry" role="listitem">
<div class="csl-left-margin">55. </div><div class="csl-right-inline">Dalhatu I, Aniekwe C, Bashorun A, et al. From <span>Paper Files</span> to <span>Web-Based Application</span> for <span>Data-Driven Monitoring</span> of <span>HIV Programs</span>: <span>Nigeria</span>’s <span>Journey</span> to a <span>National Data Repository</span> for <span>Decision-Making</span> and <span>Patient Care</span>. <em>Methods Inf Med</em>. 2023;62(03/04):130-139. doi:<a href="https://doi.org/10.1055/s-0043-1768711">10.1055/s-0043-1768711</a></div>
</div>
<div id="ref-thaiya2021adoption" class="csl-entry" role="listitem">
<div class="csl-left-margin">56. </div><div class="csl-right-inline">Thaiya MS, Julia K, Joram M, Benard M, Nambiro DA. Adoption of <span>ICT</span> to <span>Enhance Access</span> to <span>Healthcare</span> in <span>Kenya</span>. <em>IOSR-JCE</em>. 2021;23(2):45-50.</div>
</div>
<div id="ref-nsaghurwe2021one" class="csl-entry" role="listitem">
<div class="csl-left-margin">57. </div><div class="csl-right-inline">Nsaghurwe A, Dwivedi V, Ndesanjo W, et al. One country’s journey to interoperability: <span>Tanzania</span>’s experience developing and implementing a national health information exchange. <em>BMC Medical Informatics and Decision Making</em>. 2021;21(1):139. doi:<a href="https://doi.org/10.1186/s12911-021-01499-6">10.1186/s12911-021-01499-6</a></div>
</div>
<div id="ref-hapi-fhir" class="csl-entry" role="listitem">
<div class="csl-left-margin">58. </div><div class="csl-right-inline"><span>HAPI FHIR</span> - <span>The Open Source FHIR API</span> for <span>Java</span>. Accessed September 20, 2024. <a href="https://hapifhir.io/">https://hapifhir.io/</a></div>
</div>
<div id="ref-syzdykova2017opensource" class="csl-entry" role="listitem">
<div class="csl-left-margin">59. </div><div class="csl-right-inline">Syzdykova A, Malta A, Zolfo M, Diro E, Oliveira JL. Open-<span>Source Electronic Health Record Systems</span> for <span>Low-Resource Settings</span>: <span>Systematic Review</span>. <em>JMIR Medical Informatics</em>. 2017;5(4):e44. doi:<a href="https://doi.org/10.2196/medinform.8131">10.2196/medinform.8131</a></div>
</div>
<div id="ref-mehl2020open" class="csl-entry" role="listitem">
<div class="csl-left-margin">60. </div><div class="csl-right-inline">Mehl G. Open <span>Smart Register Platform</span> (<span>OpenSRP</span>). <em>OpenSRP</em>. 2020;5:42-43. Accessed January 21, 2023. <a href="https://lib.digitalsquare.io/handle/123456789/77592">https://lib.digitalsquare.io/handle/123456789/77592</a></div>
</div>
<div id="ref-open-health-stack" class="csl-entry" role="listitem">
<div class="csl-left-margin">61. </div><div class="csl-right-inline">Open <span>Health Stack</span>. Accessed September 20, 2024. <a href="https://developers.google.com/open-health-stack">https://developers.google.com/open-health-stack</a></div>
</div>
<div id="ref-summitinstitutefordevelopment2023bunda" class="csl-entry" role="listitem">
<div class="csl-left-margin">62. </div><div class="csl-right-inline">Development SI for. <span>BUNDA App</span>. May 9, 2023. Accessed January 18, 2024. <a href="https://www.sid-indonesia.org/post/bunda-app">https://www.sid-indonesia.org/post/bunda-app</a></div>
</div>
<div id="ref-kurniawan2019midwife" class="csl-entry" role="listitem">
<div class="csl-left-margin">63. </div><div class="csl-right-inline">Kurniawan K, FitriaSyah I, Jayakusuma AR, et al. Midwife service coverage, quality of work, and client health improved after deployment of an <span class="nocase">OpenSRP-driven</span> client management application in <span>Indonesia</span>. In: Atlantis Press; 2019:155-162. doi:<a href="https://doi.org/10.2991/ichs-18.2019.21">10.2991/ichs-18.2019.21</a></div>
</div>
<div id="ref-beda-emr" class="csl-entry" role="listitem">
<div class="csl-left-margin">64. </div><div class="csl-right-inline">Beda <span>EMR</span>. Accessed December 30, 2024. <a href="https://beda.software/emr">https://beda.software/emr</a></div>
</div>
<div id="ref-clickhouse" class="csl-entry" role="listitem">
<div class="csl-left-margin">65. </div><div class="csl-right-inline">ClickHouse. Clickhouse: <span>Fast Open-Source OLAP DBMS</span>. Accessed September 20, 2024. <a href="https://clickhouse.com">https://clickhouse.com</a></div>
</div>
<div id="ref-dbt" class="csl-entry" role="listitem">
<div class="csl-left-margin">66. </div><div class="csl-right-inline">Dbt. Accessed September 20, 2024. <a href="https://www.getdbt.com/index">https://www.getdbt.com/index</a></div>
</div>
<div id="ref-balch2023machine" class="csl-entry" role="listitem">
<div class="csl-left-margin">67. </div><div class="csl-right-inline">Balch JA, Ruppert MM, Loftus TJ, et al. Machine <span>Learning</span>–<span>Enabled Clinical Information Systems Using Fast Healthcare Interoperability Resources Data Standards</span>: <span>Scoping Review</span>. <em>JMIR Medical Informatics</em>. 2023;11(1):e48297. doi:<a href="https://doi.org/10.2196/48297">10.2196/48297</a></div>
</div>
<div id="ref-digitalpublicgoods" class="csl-entry" role="listitem">
<div class="csl-left-margin">68. </div><div class="csl-right-inline">Digital <span>Public Goods Alliance</span>. 2024. Accessed February 5, 2024. <a href="https://digitalpublicgoods.net/">https://digitalpublicgoods.net/</a></div>
</div>
<div id="ref-instant-openhie-v2" class="csl-entry" role="listitem">
<div class="csl-left-margin">69. </div><div class="csl-right-inline">Instant <span>OpenHIE</span> V2. Published online July 3, 2024. Accessed September 20, 2024. <a href="https://jembi.gitbook.io/instant-v2/">https://jembi.gitbook.io/instant-v2/</a></div>
</div>
<div id="ref-delussu2024survey" class="csl-entry" role="listitem">
<div class="csl-left-margin">70. </div><div class="csl-right-inline">Delussu G, Frexia F, Mascia C, et al. A survey of <span class="nocase">openEHR Clinical Data Repositories</span>. <em>International Journal of Medical Informatics</em>. 2024;191:105591. doi:<a href="https://doi.org/10.1016/j.ijmedinf.2024.105591">10.1016/j.ijmedinf.2024.105591</a></div>
</div>
<div id="ref-fhirconnectspec" class="csl-entry" role="listitem">
<div class="csl-left-margin">71. </div><div class="csl-right-inline"><span>FHIR Connect</span> specfication. Published online October 10, 2024. Accessed February 4, 2025. <a href="https://github.com/better-care/fhir-connect-mapping-spec">https://github.com/better-care/fhir-connect-mapping-spec</a></div>
</div>
<div id="ref-openfhir" class="csl-entry" role="listitem">
<div class="csl-left-margin">72. </div><div class="csl-right-inline">Welcome to <span class="nocase">openFHIR</span>’s documentation! — <span class="nocase">openFHIR</span> 0.9.3 documentation. Accessed February 4, 2025. <a href="https://open-fhir.com/documentation/index.html">https://open-fhir.com/documentation/index.html</a></div>
</div>
<div id="ref-pedrera-jimenez2023cana" class="csl-entry" role="listitem">
<div class="csl-left-margin">73. </div><div class="csl-right-inline">Pedrera-Jiménez M, García-Barrio N, Frid S, et al. Can <span>OpenEHR</span>, <span>ISO</span> 13606, and <span>HL7 FHIR Work Together</span>? <span>An Agnostic Approach</span> for the <span>Selection</span> and <span>Application</span> of <span>Electronic Health Record Standards</span> to the <span>Next-Generation Health Data Spaces</span>. <em>Journal of Medical Internet Research</em>. 2023;25(1):e48702. doi:<a href="https://doi.org/10.2196/48702">10.2196/48702</a></div>
</div>
<div id="ref-pohjonen2022norway" class="csl-entry" role="listitem">
<div class="csl-left-margin">74. </div><div class="csl-right-inline">Pohjonen H. Norway, <span>Sweden</span>, and <span>Finland</span> as forerunners in open ecosystems and <span class="nocase">openEHR</span>. In: Hovenga E, Grain H, eds. <em>Roadmap to <span>Successful Digital Health Ecosystems</span></em>. Academic Press; 2022:457-471. doi:<a href="https://doi.org/10.1016/B978-0-12-823413-6.00011-2">10.1016/B978-0-12-823413-6.00011-2</a></div>
</div>
<div id="ref-bajric2023building" class="csl-entry" role="listitem">
<div class="csl-left-margin">75. </div><div class="csl-right-inline">Bajrić S. Building a <span>Sustainable Ecosystem</span> for <span class="nocase">eHealth</span> in <span>Slovenia</span>: <span>Opportunities</span>, <span>Challenges</span>, and <span>Strategies</span>. <em>DIGITAL HEALTH</em>. 2023;9:20552076231205743. doi:<a href="https://doi.org/10.1177/20552076231205743">10.1177/20552076231205743</a></div>
</div>
<div id="ref-rsozuid-limburg2025demonstratie" class="csl-entry" role="listitem">
<div class="csl-left-margin">76. </div><div class="csl-right-inline"><em>Demonstratie Proof of Concept Regionaal Data-Ecosysteem Zuid-Limburg</em>.; 2025. Accessed February 26, 2025. <a href="https://www.youtube.com/watch?v=jT5UTLRX5VQ">https://www.youtube.com/watch?v=jT5UTLRX5VQ</a></div>
</div>
<div id="ref-wright2014open" class="csl-entry" role="listitem">
<div class="csl-left-margin">77. </div><div class="csl-right-inline">Wright SA, Druta D. Open source and standards: <span>The</span> role of open source in the dialogue between research and standardization. In: <em>2014 <span>IEEE Globecom Workshops</span> (<span>GC Wkshps</span>)</em>. IEEE; 2014:650-655. doi:<a href="https://doi.org/10.1109/GLOCOMW.2014.7063506">10.1109/GLOCOMW.2014.7063506</a></div>
</div>
<div id="ref-blind2019relationship" class="csl-entry" role="listitem">
<div class="csl-left-margin">78. </div><div class="csl-right-inline">Blind K, Böhm M. <em>The Relationship Between Open Source Software and Standard Setting</em>. Publications Office of the European Union; 2019. doi:<a href="https://doi.org/10.2760/163594">10.2760/163594</a></div>
</div>
<div id="ref-erickson2005role" class="csl-entry" role="listitem">
<div class="csl-left-margin">79. </div><div class="csl-right-inline">Erickson BJ, Langer S, Nagy P. The <span>Role</span> of <span>Open-Source Software</span> in <span>Innovation</span> and <span>Standardization</span> in <span>Radiology</span>. <em>Journal of the American College of Radiology</em>. 2005;2(11):927-931. doi:<a href="https://doi.org/10.1016/j.jacr.2005.05.004">10.1016/j.jacr.2005.05.004</a></div>
</div>
<div id="ref-nagy2007open" class="csl-entry" role="listitem">
<div class="csl-left-margin">80. </div><div class="csl-right-inline">Nagy P. Open <span>Source</span> in <span>Imaging Informatics</span>. <em>J Digit Imaging</em>. 2007;20(S1):1-10. doi:<a href="https://doi.org/10.1007/s10278-007-9056-1">10.1007/s10278-007-9056-1</a></div>
</div>
<div id="ref-simcoe2019forking" class="csl-entry" role="listitem">
<div class="csl-left-margin">81. </div><div class="csl-right-inline">Simcoe T, Watson J. Forking, <span>Fragmentation</span>, and <span>Splintering</span>. <em>Strategy Science</em>. 2019;4(4):283-297. doi:<a href="https://doi.org/10.1287/stsc.2019.0094">10.1287/stsc.2019.0094</a></div>
</div>
<div id="ref-2025sovereidn" class="csl-entry" role="listitem">
<div class="csl-left-margin">82. </div><div class="csl-right-inline">Sovereidn <span>Tech Agency</span>. February 27, 2025. Accessed February 28, 2025. <a href="https://www.sovereign.tech/">https://www.sovereign.tech/</a></div>
</div>
<div id="ref-chang2007open" class="csl-entry" role="listitem">
<div class="csl-left-margin">83. </div><div class="csl-right-inline">Chang V, Mills H, Newhouse S. From <span>Open Source</span> to long-term sustainability: <span>Review</span> of <span>Business Models</span> and <span>Case</span> studies. In:; 2007. Accessed February 28, 2025. <a href="https://eprints.soton.ac.uk/263925/">https://eprints.soton.ac.uk/263925/</a></div>
</div>
<div id="ref-nyman2013code" class="csl-entry" role="listitem">
<div class="csl-left-margin">84. </div><div class="csl-right-inline">Nyman L, Lindman J. Code <span>Forking</span>, <span>Governance</span>, and <span>Sustainability</span> in <span>Open Source Software</span>. Published online 2013.</div>
</div>
</div></section></div></main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>