<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.553">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Daniel Kapitan">
<meta name="author" content="Femke Heddema">
<meta name="author" content="Andre Dekker">
<meta name="author" content="Melle Sieswerda">
<meta name="author" content="Bart-Jan Verhoeff">
<meta name="keywords" content="OMOP, OpenEHR, FHIR, secondary use, data platform, digital platform">

<title>Data interoperability in context: the importance of open source implementations when choosing open standards</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="index_files/libs/clipboard/clipboard.min.js"></script>
<script src="index_files/libs/quarto-html/quarto.js"></script>
<script src="index_files/libs/quarto-html/popper.min.js"></script>
<script src="index_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="index_files/libs/quarto-html/anchor.min.js"></script>
<link href="index_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="index_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="index_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="index_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="index_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul class="collapse">
  <li><a href="#open-standards-are-a-necessary-but-not-sufficient-condition-for-interoperability" id="toc-open-standards-are-a-necessary-but-not-sufficient-condition-for-interoperability" class="nav-link active" data-scroll-target="#open-standards-are-a-necessary-but-not-sufficient-condition-for-interoperability">Open standards are a necessary but not sufficient condition for interoperability</a></li>
  <li><a href="#digital-platforms-require-extensibility-availibility-of-complementary-components-and-availibility-of-executable-pieces-of-software" id="toc-digital-platforms-require-extensibility-availibility-of-complementary-components-and-availibility-of-executable-pieces-of-software" class="nav-link" data-scroll-target="#digital-platforms-require-extensibility-availibility-of-complementary-components-and-availibility-of-executable-pieces-of-software">Digital platforms require extensibility, availibility of complementary components and availibility of executable pieces of software</a></li>
  <li><a href="#standardization-of-health-data-for-federated-learning" id="toc-standardization-of-health-data-for-federated-learning" class="nav-link" data-scroll-target="#standardization-of-health-data-for-federated-learning">Standardization of health data for federated learning</a></li>
  <li><a href="#health-data-standards-in-lmics" id="toc-health-data-standards-in-lmics" class="nav-link" data-scroll-target="#health-data-standards-in-lmics">Health data standards in LMICs</a></li>
  <li><a href="#conclusion" id="toc-conclusion" class="nav-link" data-scroll-target="#conclusion">Conclusion</a></li>
  <li><a href="#bibliography" id="toc-bibliography" class="nav-link" data-scroll-target="#bibliography">References</a></li>
  </ul>
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="index.pdf"><i class="bi bi-file-pdf"></i>PDF</a></li><li><a href="index.docx"><i class="bi bi-file-word"></i>MS Word</a></li></ul></div></nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Data interoperability in context: the importance of open source implementations when choosing open standards</h1>
</div>


<div class="quarto-title-meta-author">
  <div class="quarto-title-meta-heading">Authors</div>
  <div class="quarto-title-meta-heading">Affiliations</div>
  
    <div class="quarto-title-meta-contents">
    <p class="author">Daniel Kapitan <a href="mailto:daniel@kapitan.net" class="quarto-title-author-email"><i class="bi bi-envelope"></i></a> <a href="https://orcid.org/0000-0001-8979-9194" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Dutch Hospital Data
          </p>
        <p class="affiliation">
            PharmAccess Foundation
          </p>
        <p class="affiliation">
            Eindhoven University of Technology
          </p>
      </div>
    <div class="quarto-title-meta-contents">
    <p class="author">Femke Heddema </p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            PharmAccess Foundation
          </p>
      </div>
    <div class="quarto-title-meta-contents">
    <p class="author">Andre Dekker <a href="https://orcid.org/0000-0002-0422-7996" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Department of Radiation Oncology (Maastro), GROW - Research Institute for Oncology and Reproduction, Maastricht University Medical Center+
          </p>
      </div>
    <div class="quarto-title-meta-contents">
    <p class="author">Melle Sieswerda </p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Integral Cancer Registry Netherlands
          </p>
        <p class="affiliation">
            Maastricht University
          </p>
      </div>
    <div class="quarto-title-meta-contents">
    <p class="author">Bart-Jan Verhoeff </p>
  </div>
  <div class="quarto-title-meta-contents">
        <p class="affiliation">
            Expertisecentrum Zorgalgoritmen
          </p>
      </div>
  </div>

<div class="quarto-title-meta">

      
  
    
  </div>
  
<div>
  <div class="abstract">
    <div class="block-title">Abstract</div>
    <p>In response to the proposal of Tsafnat et al.&nbsp;to converge towards three open health data standards, this viewpoint provides a critical reflection on the proposed alignment of using OpenEHR, FHIR and OMOP as the default standards for clinical care and administration, data exchange and longitudinal analysis, respectively. We argue that open standards are a necessary but not sufficient condition to achieve health data interoperability. The ecosystem of open source implementations needs to be considered when choosing an appropriate standard for a given context. We discuss two specific contexts, namely standardization of i) health data for federated learning, and ii) health data sharing in low- and middle income countries (LMICs). Specific design principles, practical considerations and implementation choices for these two context are described, based on ongoing work in both areas. In the case of federated learning, we observe convergence towards OMOP and FHIR, where the two standards can effectively be used side-by-side given the availibility of mediators between the two. In the case of health information exchanges in LMICs, we see a strong convergence towards FHIR as the primary standard, with as yet limited adoption of OMOP and OpenEHR. We propose practical guidelines for context-specific adaptation of open standards.</p>
  </div>
</div>

<div>
  <div class="keywords">
    <div class="block-title">Keywords</div>
    <p>OMOP, OpenEHR, FHIR, secondary use, data platform, digital platform</p>
  </div>
</div>

</header>


<section id="open-standards-are-a-necessary-but-not-sufficient-condition-for-interoperability" class="level2">
<h2 class="anchored" data-anchor-id="open-standards-are-a-necessary-but-not-sufficient-condition-for-interoperability">Open standards are a necessary but not sufficient condition for interoperability</h2>
<p>“A paradox of health care interoperability is the existence of a large number of standards exists with significant overlap among them,” say Tsafnat et al., followed by a call to actions towards the health informatics community to put effort into establishing convergence and preventing collision <span class="citation" data-cites="tsafnat2024converge">(<a href="#ref-tsafnat2024converge" role="doc-biblioref">Tsafnat et al., 2024</a>)</span>. To do so, they propose to converge on three open standards, namely i) OpenEHR for clinical care and administration; ii) Fast Health Interoperability Resources (FHIR) for data exhange and iii) Observational Medical Outcomes Partnership Common Data Model (OMOP) for longitudinal analysis. They argue that open data standards, backed by engaged communities, hold an advantage over proprietary ones and therefore should be chosen as the steppingstones towards achieving true interoperability.</p>
<p>While we support their high-level rationale and intention, we feel their proposed trichotomy does not do justice to details that are crucial in real-world implementations. This viewpoint provides a critical reflection on their proposed framework in three parts. First, we reflect on salient differences between the three open standards from the perspective of the notion of openness of digital platforms <span class="citation" data-cites="dereuver2018digital">(<a href="#ref-dereuver2018digital" role="doc-biblioref">de Reuver et al., 2018</a>)</span> and the paradox of open <span class="citation" data-cites="keller2021paradox">(<a href="#ref-keller2021paradox" role="doc-biblioref">Keller and Tarkowski, 2021</a>)</span>. Subsequently, we present our findings in designing and implementing health data platforms in two specific contexts, namely i) platforms for federated learning on shared health data in high income countries; and ii) health data platforms for low and middle income countries (LMICs). We conclude with practical guidelines for context-specific adaptation of open standards.</p>
</section>
<section id="digital-platforms-require-extensibility-availibility-of-complementary-components-and-availibility-of-executable-pieces-of-software" class="level2">
<h2 class="anchored" data-anchor-id="digital-platforms-require-extensibility-availibility-of-complementary-components-and-availibility-of-executable-pieces-of-software">Digital platforms require extensibility, availibility of complementary components and availibility of executable pieces of software</h2>
<p>Besides the paradox of interoperability put forward by Tsafnat et al., we argue that open standards are a necessary, but not sufficient condition for convergence of health data standarization. Open source implementations of components, software development kits etc. constitute another necessary condition for establishing a flourishing health data sharing platform and associated ecosystem for any given context, be it regional, international or within a specific sub-domain like pandemic preparedness. Research on digital platforms underline the importance of the platform openness, not only in term of open standards, but also in term of extensibility of the code base, availibility of complements to the core technical platform (in our case the data standard itself) and availibility of executable pieces of software <span class="citation" data-cites="dereuver2018digital">(<a href="#ref-dereuver2018digital" role="doc-biblioref">de Reuver et al., 2018</a>)</span>. Only when the majority of these aspects of digital platforms are fullfilled can we resonably expect that the platform will indeed be longlived.</p>
<p>In what they call the paradox of open, Keller and Tarkowsi argue that this conventional approach of open standards and open source flourish under two types of conditions <span class="citation" data-cites="keller2021paradox">(<a href="#ref-keller2021paradox" role="doc-biblioref">Keller and Tarkowski, 2021</a>)</span>. First, projects where many people contribute to the creation of a common resource have proven succesful. “This is the story of Wikipedia, OpenStreetMap, Blender.org, and the countless free software projects that provide much of the internet’s infrastructure.” <span class="citation" data-cites="keller2021paradox">(<a href="#ref-keller2021paradox" role="doc-biblioref">Keller and Tarkowski, 2021</a>)</span> Indeed, Tsafnat et al.&nbsp;have explicitly taken into account that “an engaged and vibrant community is a major advantage for the longevity of the data standards it uses,” which has informed their proposal to converge towards OMOP, FHIR en OpenEHR. However, the emphasis on open source implementations is somewhat overlooked. This point is only mentioned in passing and indirectly, when Tsafnat et al.&nbsp;reference work done by Reynolds and Wyatt who already argued in 2011 “… for the superiority of open source licensing to promote safer, more effective health care information systems. We claim that open source licensing in health care information systems is essential to rational procurement strategy” <span class="citation" data-cites="reynolds2011open">(<a href="#ref-reynolds2011open" role="doc-biblioref">Reynolds and Wyatt, 2011</a>)</span>. We believe that a realistic assessment of the current position of an open standard within the wider context of availability of complementary components and open source implementations is equally important when choosing which standard to adopt.</p>
<p>This point is related to the second condition put forward by Keller and Tarkoswki, namely that the conventional open approach has proven fruitful when “opening up” is the result of external incentives or requirements, rather than voluntary actions. “This is the story of publicly-funded knowledge production like Open Access academic publications, cultural heritage collections in the Public Domain, Open Educational Resources (OER), and Open Government data.” <span class="citation" data-cites="keller2021paradox">(<a href="#ref-keller2021paradox" role="doc-biblioref">Keller and Tarkowski, 2021</a>)</span> A canonical example is the birth of the GSM standard, which was mandated by European legislation.<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a> Reflecting on this perspective on openness, we observe a salient difference between FHIR vis-a-vis OpenEHR and OMOP, namely that the former is the only one that has been mandated (or at least strongly recommended) in some jurisdictions. In the US, the Office of the National Coordinator for Health Information Technology (ONC) and the Centers for Medicare and Medicaid Services (CMS) has introduced a steady stream of new regulations, criteria, and deadlines in Health IT that has resulted in significant adoption of FHIR <span class="citation" data-cites="firely2023fhir">(<a href="#ref-firely2023fhir" role="doc-biblioref">Firely, 2023</a>)</span>. In India, the open Health Claims Exchange protocol specification - which is based on FHIR - has been mandated by the Indian government as the standard for e-claims handling <span class="citation" data-cites="india2020national hcx">(<a href="#ref-hcx" role="doc-biblioref"><span>“<span>HCX Protocol</span> v0.9,”</span> 2023</a>; <a href="#ref-india2020national" role="doc-biblioref"><em>National <span>Digital Health Mission</span></em>, 2020</a>)</span>. The African Union recommends all new implementations and digital health system improvements use FHIR as the primary mechanism for data exchange <span class="citation" data-cites="tilahun2023african">(<a href="#ref-tilahun2023african" role="doc-biblioref">Tilahun et al., 2023</a>)</span>, but doesn’t say anything about the use of, for example, OpenEHR for administrative point-of-service systems.</p>
<p>These external incentives have resulted in a large boost in both commercial and open source development activities in the FHIR ecosystem. Illustrative of this is the speed with which the Bulk FHIR API has been defined and implemented in almost all major implementations <span class="citation" data-cites="mandl2020push jones2021landscape">(<a href="#ref-jones2021landscape" role="doc-biblioref">Jones et al., 2021</a>; <a href="#ref-mandl2020push" role="doc-biblioref">Mandl et al., 2020</a>)</span>, and the the SQL-on-FHIR specification to make large-scale analysis of FHIR data accessible to a larger audience and portable between systems.<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a> It has also led to more people voluntarily contributing to FHIR-related open source projects, which has resulted in a wide offering of FHIR components across major technology stacks (Java, Python, .NET), thereby strengthening the first condition. By comparison, OMOP and OpenEHR have not yet profited from external incentives to spur the adoption and thereby growing the ecosystem beyond a certain critical mass. To illustrate this, a search on GitHub on “FHIR” yields 8.2 thousand results, “OMOP or OHDSI” one thousand results, and “OpenEHR” returns 400 results. A quick-scan of the available open source components listed on the website of the three governing bodies HL7, OHDSI and OpenEHR, indicates that the ecosystem of FHIR and OMOP have a significantly larger offering of extensible and complementary open source components than OpenEHR.<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a></p>
<p>Hence, we stress that beyond the evaluating the instrinic structure of an open standard and the community that supports the standard, we need to take into account the wider ecosystem of open source implementations and availibility of complementary components. From this wider perspective of the whole ecosystem surrounding the three standards, FHIR stands out as having the most diverse and rich ecosystem because it has been mandated in certain jurisdictions. This is relevant when comparing these standards in real-world implementations. We now turn to two specific use cases c.q. contexts where these considerations are at play.</p>
</section>
<section id="standardization-of-health-data-for-federated-learning" class="level2">
<h2 class="anchored" data-anchor-id="standardization-of-health-data-for-federated-learning">Standardization of health data for federated learning</h2>
<p>The current fragmentation in health data is one of the major barriers towards leveraging the potential medical data for machine learning (ML). Without access to sufficient data, ML will be limited in its application to health improvement efforts and, ultimately, from making the transition from research to clinical practice. High quality health data, obtained from a research setting or a real-world clinical practice setting, is hard to obtain, because health data is highly sensitive and its usage is tightly regulated.</p>
<p>Federated learning (FL) is a learning paradigm that aims to address these issues of data governance and privacy by training algorithms collaboratively without moving (copying) the data itself <span class="citation" data-cites="rieke2020future teo2024federated">(<a href="#ref-rieke2020future" role="doc-biblioref">Rieke et al., 2020</a>; <a href="#ref-teo2024federated" role="doc-biblioref">Teo et al., 2024</a>)</span>. Based on ongoing work with the PLUGIN healthcare consortium (<a href="https://plugin.healthcare">https://plugin.healthcare</a>, in Dutch) we have detailed an architecture for FL for secondary use of health data for hospitals in the Netherlands. Starting point for this implementation are the National Health Data Infrastructure agreements for research, policy and innovation for the Dutch healthcare sector, which have been adopted at the beginning of 2024 <span class="citation" data-cites="healthri2024agreements">(<a href="#ref-healthri2024agreements" role="doc-biblioref">Health-RI, 2024</a>)</span>. <a href="#fig-healthri-architecture" class="quarto-xref">Figure&nbsp;1</a> shows a high level overview of the platform, which comprises three areas (multiple use, applications and generic features) and a total of 26 functional components (for details please refer to <span class="citation" data-cites="healthri2024agreements">(<a href="#ref-healthri2024agreements" role="doc-biblioref">Health-RI, 2024</a>)</span>). One of the prerequisites of this architecture is that organizations that participate in a federation of ‘data stations’ use the same common data model to make the data Findable, Accessible, Interoperable and Resusable (FAIR). These FAIR data stations comprise components 7, 8 and 9 in <a href="#fig-healthri-architecture" class="quarto-xref">Figure&nbsp;1</a>, i.e.&nbsp;the data, metadata and APIs, respectively, through which this the data station can be accessed and used.</p>
<div id="fig-healthri-architecture" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-fig" id="fig-healthri-architecture-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: Reference architecture for the Dutch health data infrastructure for research and innovation <span class="citation" data-cites="healthri2024agreements">(<a href="#ref-healthri2024agreements" role="doc-biblioref">Health-RI, 2024</a>)</span>
</figcaption>
<div aria-describedby="fig-healthri-architecture-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="health-ri-architecture.png" class="img-fluid figure-img">
</div>
</figure>
</div>
<p>Following the line of reasoning of Tsafnat et al., OMOP would be the go-to standard for storing the longitudinal data in each of the data stations. Indeed, by now there are quite a few reports of real-world implementations of federated learning networks based on the OHDSI-OMOP stack, including a global infrastructure with 22 centres for COVID19 prediction models <span class="citation" data-cites="khalid2021standardized">(<a href="#ref-khalid2021standardized" role="doc-biblioref">Khalid et al., 2021</a>)</span>, FeederNet in South Korea with 57 participating hospitals <span class="citation" data-cites="lee2022feedernet">(<a href="#ref-lee2022feedernet" role="doc-biblioref">Lee et al., 2022</a>)</span>, Dutch multi-cohort dementia research with 9 centres <span class="citation" data-cites="mateus2024data">(<a href="#ref-mateus2024data" role="doc-biblioref">Mateus et al., 2024</a>)</span>, the European severe heterogeneous asthma research collaboration <span class="citation" data-cites="kroes2022blueprint">(<a href="#ref-kroes2022blueprint" role="doc-biblioref">Kroes et al., 2022</a>)</span> and the recently initiated Belgian Federated Health Innovation Network (FHIN) <a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a>.</p>
<p>For the PLUGIN project, however, we choose to adopt FHIR as the common data model because of its practicality and extensibility to be used in a Python-based data science stack, provenance of RESTful APIs out-of-the-box to facilitate easy integration with the container-based vantage6 FL framework, and the support of many healthcare terminologies and flexibility through the profiling mechanims <span class="citation" data-cites="choudhury2020personal smits2022improved">(<a href="#ref-choudhury2020personal" role="doc-biblioref">Choudhury et al., 2020</a>; <a href="#ref-smits2022improved" role="doc-biblioref">Smits et al., 2022</a>)</span>. Increasingly, other projects have reported the use of FHIR for persistent, longitudinal storage for FL. The CODA platform, which aims to implement a similar FL infrastructure in Canada, compared OMOP and FHIR and chose the latter as it has been found to support more granular mappings required for analytics <span class="citation" data-cites="mullie2023coda">(<a href="#ref-mullie2023coda" role="doc-biblioref">Mullie et al., 2023</a>)</span>. The fair4health project has implemented also based on FHIR, using their own open source framework for the federated learning infrastructure itself <span class="citation" data-cites="sinaci2024privacypreserving">(<a href="#ref-sinaci2024privacypreserving" role="doc-biblioref">Sinaci et al., 2024</a>)</span>.</p>
<p>Given that conceptually OMOP can be viewed as a strict subset of FHIR, hybrid solutions using OMOP and FHIR combined have also been reported, such as the German KETOS platform <span class="citation" data-cites="gruendner2019ketos">(<a href="#ref-gruendner2019ketos" role="doc-biblioref">Gruendner et al., 2019</a>)</span>, and the preliminary findings from the European GenoMed4All project which aims to connect clinical and -omics data <span class="citation" data-cites="cremonesi2023need">(<a href="#ref-cremonesi2023need" role="doc-biblioref">Cremonesi et al., 2023</a>)</span>. A collaboration of 10 university hospitals in Germany have shown that standardized ETL-processing from FHIR into OMOP can achieve 99% conformance <span class="citation" data-cites="peng2023etlprocess">(<a href="#ref-peng2023etlprocess" role="doc-biblioref">Peng et al., 2023</a>)</span>, which confirms the feasiblity of the solution pattern where FHIR acts as an intermediate sharing standard through which data from (legacy) systems are extracted and made available for reuse in a common data model. One could argue that the distiction between FHIR amd OMOP becomes less relevant if data can be effectively stored in either standard. We are hopeful that initiatives like OMOP-on-FHIR indeed will foster convergence rather than collision between these two standards.<a href="#fn5" class="footnote-ref" id="fnref5" role="doc-noteref"><sup>5</sup></a></p>
<p>In the case of PLUGIN, another important consideration for choosing FHIR over OMAP is, that from a data architecture perspective, the mechanism of FHIR Profiles can be tied to principle of late binding commonly applied in data lake/warehouse architectures: allow ingest of widely different sources, and gradually add more constraints and validations as you move closer to a specific use case. If machine learning is the primary objective for secondary use, we want to be able to cast a wider net of relevant data, rather than being to restrictive when ingesting the data at the start of processing pipeline. Late binding in data warehousing is a design philosophy where data transformation and schema enforcement are deferred as late as possible in the data processing pipeline, sometimes even until query time. This approach contrasts with early binding, where data is transformed and structured as it is ingested into the data warehouse. This principle is visualized as concentric circles (<a href="#fig-late-binding" class="quarto-xref">Figure&nbsp;2</a>). The advantages of this design is that it allows for greater flexibility. During the initial ingest of the data, we only require the data to conform to the minimal syntactic standard defined by the base FHIR version (R4 in the diagram). As the data is processed, more strict checks and constrains are applied, whereby ultimately different profiles can co-exists next to one another (the two most inner circles), within a larger circle with fewer strictions. This approach does not support the extension mechanism of FHIR, so we need to be cautious if we decide to use that.</p>
<div id="fig-late-binding" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-fig" id="fig-late-binding-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2: Principle of late binding with FHIR profiling mechanism
</figcaption>
<div aria-describedby="fig-late-binding-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="late-binding.png" class="img-fluid figure-img">
</div>
</figure>
</div>
<p>We found that this principle of late binding also allows flexible and efficient implementations of the data stations that make use of the current best practices of the a lakehouse architecture of <span class="citation" data-cites="hai2023data harby2022data harby2024data">(<a href="#ref-hai2023data" role="doc-biblioref">Hai et al., 2023</a>; <a href="#ref-harby2024data" role="doc-biblioref">Harby and Zulkernine, 2024</a>, <a href="#ref-harby2022data" role="doc-biblioref">2022</a>)</span> and the composable data stack <span class="citation" data-cites="pedreira2023composable">(<a href="#ref-pedreira2023composable" role="doc-biblioref">Pedreira et al., 2023</a>)</span>. Lakehouses typically have a zonal architecture that follow the Extract-Load-Transform pattern (ELT) where data is ingested from the source systems in bulk (E), delivered to storage with aligned schemas (L) and transformed into a format ready for analysis (T) <span class="citation" data-cites="hai2023data">(<a href="#ref-hai2023data" role="doc-biblioref">Hai et al., 2023</a>)</span>. The discerning characteristic of the lakehouse architecture is its foundation on low-cost and directly-accessible storage that also provides traditional database management and performance features such as ACID transactions, data versioning, auditing, indexing, caching, and query optimization <span class="citation" data-cites="armbrust2021lakehouse">(<a href="#ref-armbrust2021lakehouse" role="doc-biblioref">Armbrust et al., 2021</a>)</span>. Lakehouses thus combine the key benefits of data lakes and data warehouses: low-cost storage in an open format accessible by a variety of systems from the former, and powerful management and optimization features from the latter. By explicitly aligining the mechanism of FHIR Profiles with this design pattern of a data lakehouse enables us to use complementary standards and open source components, most notably Apache Arrow as the standard columnar in-memory format with RPC-based data movement; Apache Parquet as the standard columnar on-disk format; and Apache Iceberg as the open table format.<a href="#fn6" class="footnote-ref" id="fnref6" role="doc-noteref"><sup>6</sup></a></p>
<p>The main disadvantage in using FHIR in this way pertains to the need for upgrading the whole ELT pipeline when upgrading to a new primary FHIR version, for example R5. However, we expect that the development time required to upgrade FHIR versions is significantly less than the initial migration to FHIR.</p>
<p>[TO DO: add closing comments for this section]</p>
</section>
<section id="health-data-standards-in-lmics" class="level2">
<h2 class="anchored" data-anchor-id="health-data-standards-in-lmics">Health data standards in LMICs</h2>
<p>It is a widely held belief that digital technologies have an important role to play in strengthening health systems in LMICs. Yet, also here the current fragmentation of health data stands in the way of scaling up digital health programmes beyond project-centric, vertical solutions into sustainable health information exchanges. Mehl et al.&nbsp;have called for convergence to open standards, similar to Tsafnat et al., but additionally stress the need for open source technologies (our main argument of this paper), open content (representations of public health, health system or clinical knowledge to guide implementations) and open architectures (reusable enterprise architecture patterns for health systems) <span class="citation" data-cites="mehl2023fullstac">(<a href="#ref-mehl2023fullstac" role="doc-biblioref">Mehl et al., 2023</a>)</span>. As for the open architecture, we see a convergence towards the OpenHIE framework <span class="citation" data-cites="openhie">(<a href="#ref-openhie" role="doc-biblioref"><span>“<span>OpenHIE Framework</span> v5.0,”</span> 2022</a>)</span>, which has been adopted by many sub-Saharan African countries as the architectural blueprint for implementing nation-wide health information exchanges (HIE) <span class="citation" data-cites="mamuye2022health">(<a href="#ref-mamuye2022health" role="doc-biblioref">Mamuye et al., 2022</a>)</span>, including Nigeria <span class="citation" data-cites="dalhatu2023paper">(<a href="#ref-dalhatu2023paper" role="doc-biblioref">Dalhatu et al., 2023</a>)</span>, Kenya <span class="citation" data-cites="thaiya2021adoption">(<a href="#ref-thaiya2021adoption" role="doc-biblioref">Thaiya et al., 2021</a>)</span> and Tanzania <span class="citation" data-cites="nsaghurwe2021one">(<a href="#ref-nsaghurwe2021one" role="doc-biblioref">Nsaghurwe et al., 2021</a>)</span>. <a href="#fig-openhie" class="quarto-xref">Figure&nbsp;3</a> shows an overview of the OpenHIE architecture.</p>
<div id="fig-openhie" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-fig" id="fig-openhie-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3: OpenHIE architecture showing the Point of Service systems (black), the Interoperability Layer (green) and the Component Layer (blue).
</figcaption>
<div aria-describedby="fig-openhie-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="openhie.png" class="img-fluid figure-img">
</div>
</figure>
</div>
<p>While the OpenHIE specification is agnostic to which data standard should be used, in practice the digital health community in LMICs have <em>de facto</em> converged towards FHIR as main standard, in line with the proposal by Tsafnat et al.&nbsp;for health information exchange. To illustrate this point, consider the OpenHIM Platform architecture, which is currently the largest open source implementation of the OpenHIE specification <a href="#fig-openhim-platform" class="quarto-xref">Figure&nbsp;4</a>. Clients (Point-of-Service systems) can initiate various workflows to submit or query patient data. The Shared Health Record (SHR) acts as the core transactional system for the health information exchange, which in this case is realized with the HAPI FHIR server<a href="#fn7" class="footnote-ref" id="fnref7" role="doc-noteref"><sup>7</sup></a>, being one of the most widely used open source implementations ().</p>
<div id="fig-openhim-platform" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-fig" id="fig-openhim-platform-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4: OpenHIM Platform Architecture, illustrating the use of FHIR-based workflows between the components as specified in OpenHIE. CR: Client Registry. IOL: Interoperability Layer. MPI: Master Patient Index. SHR: Shared Health Record. Image taken from <a href="https://jembi.gitbook.io/">https://jembi.gitbook.io/</a>.
</figcaption>
<div aria-describedby="fig-openhim-platform-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="openhim-platform.png" class="img-fluid figure-img">
</div>
</figure>
</div>
<p>Looking at the Point of Service systems, we see that as of today OpenEHR does not play any role of significance as the standard for clinical administration. Within the context of LMICs, the largest open source EHR implementations are based on proprietary data models, and it is unlikely this will change any time soon <span class="citation" data-cites="syzdykova2017opensource">(<a href="#ref-syzdykova2017opensource" role="doc-biblioref">Syzdykova et al., 2017</a>)</span>. Instead, we see that the FHIR-native OpenSRP framework <span class="citation" data-cites="mehl2020open">(<a href="#ref-mehl2020open" role="doc-biblioref">Mehl, 2020</a>)</span> being used more and more, where Android apps are used for clinical administration by health professionals (<a href="#fig-opensrp" class="quarto-xref">Figure&nbsp;5</a>). OpenSRP has been deployed in 14 countries targeting various patient populations, amongst which a reference implementation of the WHO antenatal and neonatal care guidelines for midwives in Lombok, Indonesia <span class="citation" data-cites="summitinstitutefordevelopment2023bunda kurniawan2019midwife">(<a href="#ref-kurniawan2019midwife" role="doc-biblioref">Kurniawan et al., 2019</a>; <a href="#ref-summitinstitutefordevelopment2023bunda" role="doc-biblioref">Summit Institute for Development, 2023</a>)</span>. This solution design is particularly useful for mid-size and smaller healthcare facilities, which are often resource constrained, lacking basic IT infrastructure to deploy as full-blown electronic medical record system. Hence, by necessity, the FHIR-based SHR functions as the clinical administration system of record and as the hub for information exchange.</p>
<div id="fig-opensrp" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-fig" id="fig-opensrp-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5: Overview of OpenSRP2 open source framework for building clinical administration apps. HIS: health information systems. Image source: <a href="https://docs.opensrp.io/">https://docs.opensrp.io/</a>.
</figcaption>
<div aria-describedby="fig-opensrp-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="opensrp.png" class="img-fluid figure-img">
</div>
</figure>
</div>
<p>Finally, regarding longitudinal data analysis we also see a convergence towards FHIR as the primary standard. As is the case of federated learning, the choice for FHIR to implement datawarehouse and analytic platforms is the preferred method due to the widespread availibility of complementary open source technologies. FHIR-specific technologies such as Bulk FHIR data access and SQL-on-FHIR mentioned earlier, allow the FHIR ecosystem to be used, complemented and integrated with generic open source datawarehousing technologies such as Clickhouse<a href="#fn8" class="footnote-ref" id="fnref8" role="doc-noteref"><sup>8</sup></a> and dbt.<a href="#fn9" class="footnote-ref" id="fnref9" role="doc-noteref"><sup>9</sup></a></p>
<p>To summarize, we see that in the context of LMICs, the standardization of the three domains put forward by Tsafnat merge into one. The SHR, as the key component within the OpenHIE specification, serves as the back-end of the system-of-record and provides a transactional, persistent storage engine for information exchange. Downstream longitudinal data stores continu to use FHIR as the common data model for analytical purposes. One could argue that it may even be advantageous to converge to just one standard, thereby reducing complexity and cost of the total system. Thanks to the increasing availibility of open source implementations as digital public goods <span class="citation" data-cites="digitalpublicgoods">(<a href="#ref-digitalpublicgoods" role="doc-biblioref"><span>“Digital <span>Public Goods Alliance</span>,”</span> 2024</a>)</span> and integration projects such as Instant OpenHIE<a href="#fn10" class="footnote-ref" id="fnref10" role="doc-noteref"><sup>10</sup></a> that we have a chance to move the needle in health data standardization for LMICs.</p>
</section>
<section id="conclusion" class="level2">
<h2 class="anchored" data-anchor-id="conclusion">Conclusion</h2>
<p>We agree with Tsafnat et al.&nbsp;that there is a dire need to converge to open data standards in healtcare, and support the proposal to focus on OpenEHR, FHIR and OMOP developments in healthcare informatics going forward. However, open standards are a necessary but not sufficient condition for convergence of health data standardization. The availiblity of open source implementations and complementary technologies are as important when choosing which open standard to use. Furthermore, we find that the proposed trichotomy is not always relevant, as we have shown in the case of federated learning en health information exchange in LMICs. As an alternative, we find that the full-STAC approach described by Mehl et al.&nbsp;more comprehensive.</p>
<p>InstantHIE ()to deploy a swarm of containers to provision all relevant components process Gebreslassie et al., 2023)</p>
</section>
<section id="bibliography" class="level1 unnumbered">


</section>


<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-armbrust2021lakehouse" class="csl-entry" role="listitem">
Armbrust, M., Ghodsi, A., Xin, R., Zaharia, M., 2021. Lakehouse: <span>A New Generation</span> of <span>Open Platforms</span> that <span>Unify Data Warehousing</span> and <span>Advanced Analytics</span>, in: 11th <span>Annual Conference</span> on <span>Innovative Data Systems Research</span> (<span>CIDR</span> ’21). p. 8.
</div>
<div id="ref-choudhury2020personal" class="csl-entry" role="listitem">
Choudhury, A., van Soest, J., Nayak, S., Dekker, A., 2020. Personal <span>Health Train</span> on <span>FHIR</span>: <span>A Privacy Preserving Federated Approach</span> for <span>Analyzing FAIR Data</span> in <span>Healthcare</span>, in: Bhattacharjee, A., Borgohain, S.Kr., Soni, B., Verma, G., Gao, X.-Z. (Eds.), Machine <span>Learning</span>, <span>Image Processing</span>, <span>Network Security</span> and <span>Data Sciences</span>, Communications in <span>Computer</span> and <span>Information Science</span>. Springer, Singapore, pp. 85–95. <a href="https://doi.org/10.1007/978-981-15-6315-7_7">https://doi.org/10.1007/978-981-15-6315-7_7</a>
</div>
<div id="ref-cremonesi2023need" class="csl-entry" role="listitem">
Cremonesi, F., Planat, V., Kalokyri, V., Kondylakis, H., Sanavia, T., Miguel Mateos Resinas, V., Singh, B., Uribe, S., 2023. The need for multimodal health data modeling: <span>A</span> practical approach for a federated-learning healthcare platform. Journal of Biomedical Informatics 141, 104338. <a href="https://doi.org/10.1016/j.jbi.2023.104338">https://doi.org/10.1016/j.jbi.2023.104338</a>
</div>
<div id="ref-dalhatu2023paper" class="csl-entry" role="listitem">
Dalhatu, I., Aniekwe, C., Bashorun, A., Abdulkadir, A., Dirlikov, E., Ohakanu, S., Adedokun, O., Oladipo, A., Jahun, I., Murie, L., Yoon, S., Abdu-Aguye, M.G., Sylvanus, A., Indyer, S., Abbas, I., Bello, M., Nalda, N., Alagi, M., Odafe, S., Adebajo, S., Ogorry, O., Akpu, M., Okoye, I., Kakanfo, K., Onovo, A.A., Ashefor, G., Nzelu, C., Ikpeazu, A., Aliyu, G., Ellerbrock, T., Boyd, M., Stafford, K.A., Swaminathan, M., 2023. From <span>Paper Files</span> to <span>Web-Based Application</span> for <span>Data-Driven Monitoring</span> of <span>HIV Programs</span>: <span>Nigeria</span>’s <span>Journey</span> to a <span>National Data Repository</span> for <span>Decision-Making</span> and <span>Patient Care</span>. Methods of Information in Medicine 62, 130–139. <a href="https://doi.org/10.1055/s-0043-1768711">https://doi.org/10.1055/s-0043-1768711</a>
</div>
<div id="ref-dereuver2018digital" class="csl-entry" role="listitem">
de Reuver, M., Sørensen, C., Basole, R.C., 2018. The <span>Digital Platform</span>: <span>A Research Agenda</span>. Journal of Information Technology 33, 124–135. <a href="https://doi.org/10.1057/s41265-016-0033-3">https://doi.org/10.1057/s41265-016-0033-3</a>
</div>
<div id="ref-digitalpublicgoods" class="csl-entry" role="listitem">
Digital <span>Public Goods Alliance</span>, 2024. Digital Public Goods Alliance - Promoting digital public goods to create a more equitable world.
</div>
<div id="ref-firely2023fhir" class="csl-entry" role="listitem">
Firely, 2023. <span>FHIR</span> in <span>US</span> healthcare regulations.
</div>
<div id="ref-gruendner2019ketos" class="csl-entry" role="listitem">
Gruendner, J., Schwachhofer, T., Sippl, P., Wolf, N., Erpenbeck, M., Gulden, C., Kapsner, L.A., Zierk, J., Mate, S., Stürzl, M., Croner, R., Prokosch, H.-U., Toddenroth, D., 2019. <span>KETOS</span>: <span>Clinical</span> decision support and machine learning as a service – <span>A</span> training and deployment platform based on <span>Docker</span>, <span>OMOP-CDM</span>, and <span>FHIR Web Services</span>. PLOS ONE 14, e0223010. <a href="https://doi.org/10.1371/journal.pone.0223010">https://doi.org/10.1371/journal.pone.0223010</a>
</div>
<div id="ref-hai2023data" class="csl-entry" role="listitem">
Hai, R., Koutras, C., Quix, C., Jarke, M., 2023. Data <span>Lakes</span>: <span>A Survey</span> of <span>Functions</span> and <span>Systems</span>. IEEE Transactions on Knowledge and Data Engineering 35, 12571–12590. <a href="https://doi.org/10.1109/TKDE.2023.3270101">https://doi.org/10.1109/TKDE.2023.3270101</a>
</div>
<div id="ref-harby2024data" class="csl-entry" role="listitem">
Harby, A.A., Zulkernine, F., 2024. Data <span>Lakehouse</span>: <span>A Survey</span> and <span>Experimental Study</span>. <a href="https://doi.org/10.2139/ssrn.4765588">https://doi.org/10.2139/ssrn.4765588</a>
</div>
<div id="ref-harby2022data" class="csl-entry" role="listitem">
Harby, A.A., Zulkernine, F., 2022. From <span>Data Warehouse</span> to <span>Lakehouse</span>: <span>A Comparative Review</span>, in: 2022 <span>IEEE International Conference</span> on <span>Big Data</span> (<span>Big Data</span>). IEEE, Osaka, Japan, pp. 389–395. <a href="https://doi.org/10.1109/BigData55660.2022.10020719">https://doi.org/10.1109/BigData55660.2022.10020719</a>
</div>
<div id="ref-hcx" class="csl-entry" role="listitem">
<span>HCX Protocol</span> v0.9, 2023.
</div>
<div id="ref-healthri2024agreements" class="csl-entry" role="listitem">
Health-RI, 2024. Agreements on the <span>National Health Data Infrastructure</span> for <span>Research</span>, <span>Policy</span> and <span>Innovation</span> - <span class="nocase">Health-RI Nationale Gezondheidsdata-infrastructuur</span> - <span>Confluence</span>.
</div>
<div id="ref-jones2021landscape" class="csl-entry" role="listitem">
Jones, J., Gottlieb, D., Mandel, J.C., Ignatov, V., Ellis, A., Kubick, W., Mandl, K.D., 2021. A landscape survey of planned <span>SMART</span>/<span>HL7</span> bulk <span>FHIR</span> data access <span>API</span> implementations and tools. Journal of the American Medical Informatics Association 28, 1284–1287. <a href="https://doi.org/10.1093/jamia/ocab028">https://doi.org/10.1093/jamia/ocab028</a>
</div>
<div id="ref-keller2021paradox" class="csl-entry" role="listitem">
Keller, P., Tarkowski, A., 2021. The <span>Paradox</span> of <span>Open</span>. Open Future.
</div>
<div id="ref-khalid2021standardized" class="csl-entry" role="listitem">
Khalid, S., Yang, C., Blacketer, C., Duarte-Salles, T., Fernández-Bertolín, S., Kim, C., Park, R.W., Park, J., Schuemie, M.J., Sena, A.G., Suchard, M.A., You, S.C., Rijnbeek, P.R., Reps, J.M., 2021. A standardized analytics pipeline for reliable and rapid development and validation of prediction models using observational health data. Computer Methods and Programs in Biomedicine 211, 106394. <a href="https://doi.org/10.1016/j.cmpb.2021.106394">https://doi.org/10.1016/j.cmpb.2021.106394</a>
</div>
<div id="ref-kroes2022blueprint" class="csl-entry" role="listitem">
Kroes, J.A., Bansal, A.T., Berret, E., Christian, N., Kremer, A., Alloni, A., Gabetta, M., Marshall, C., Wagers, S., Djukanovic, R., Porsbjerg, C., Hamerlijnck, D., Fulton, O., Brinke, A. ten, Bel, E.H., Sont, J.K., 2022. Blueprint for harmonising unstandardised disease registries to allow federated data analysis: Prepare for the future. ERJ Open Research 8. <a href="https://doi.org/10.1183/23120541.00168-2022">https://doi.org/10.1183/23120541.00168-2022</a>
</div>
<div id="ref-kurniawan2019midwife" class="csl-entry" role="listitem">
Kurniawan, K., FitriaSyah, I., Jayakusuma, A.R., Armis, R.A., Lubis, Y., Haryono, M.A., Harefa, B., Shankar, A., 2019. Midwife service coverage, quality of work, and client health improved after deployment of an <span class="nocase">OpenSRP-driven</span> client management application in <span>Indonesia</span>, in: 5th <span>International Conference</span> on <span>Health Sciences</span> (<span>ICHS</span> 2018). Atlantis Press, pp. 155–162. <a href="https://doi.org/10.2991/ichs-18.2019.21">https://doi.org/10.2991/ichs-18.2019.21</a>
</div>
<div id="ref-lee2022feedernet" class="csl-entry" role="listitem">
Lee, S., Kim, C., Chang, J., Park, R.W., 2022. <span>FeederNet</span> (<span>Federated E-Health Big Data</span> for <span>Evidence Renovation Network</span>) platform in <span>Korea</span> – <span>OHDSI</span>.
</div>
<div id="ref-mamuye2022health" class="csl-entry" role="listitem">
Mamuye, A.L., Yilma, T.M., Abdulwahab, A., Broomhead, S., Zondo, P., Kyeng, M., Maeda, J., Abdulaziz, M., Wuhib, T., Tilahun, B.C., 2022. Health information exchange policy and standards for digital health systems in africa: <span>A</span> systematic review. PLOS Digital Health 1, e0000118. <a href="https://doi.org/10.1371/journal.pdig.0000118">https://doi.org/10.1371/journal.pdig.0000118</a>
</div>
<div id="ref-mandl2020push" class="csl-entry" role="listitem">
Mandl, K.D., Gottlieb, D., Mandel, J.C., Ignatov, V., Sayeed, R., Grieve, G., Jones, J., Ellis, A., Culbertson, A., 2020. Push <span>Button Population Health</span>: <span>The SMART</span>/<span>HL7 FHIR Bulk Data Access Application Programming Interface</span>. npj Digital Medicine 3, 1–9. <a href="https://doi.org/10.1038/s41746-020-00358-4">https://doi.org/10.1038/s41746-020-00358-4</a>
</div>
<div id="ref-mateus2024data" class="csl-entry" role="listitem">
Mateus, P., Moonen, J., Beran, M., Jaarsma, E., van der Landen, S.M., Heuvelink, J., Birhanu, M., Harms, A.G.J., Bron, E., Wolters, F.J., Cats, D., Mei, H., Oomens, J., Jansen, W., Schram, M.T., Dekker, A., Bermejo, I., 2024. Data harmonization and federated learning for multi-cohort dementia research using the <span>OMOP</span> common data model: <span>A Netherlands</span> consortium of dementia cohorts case study. Journal of Biomedical Informatics 155, 104661. <a href="https://doi.org/10.1016/j.jbi.2024.104661">https://doi.org/10.1016/j.jbi.2024.104661</a>
</div>
<div id="ref-mehl2020open" class="csl-entry" role="listitem">
Mehl, G., 2020. Open <span>Smart Register Platform</span> (<span>OpenSRP</span>). mHealth Compendium 5, 42–43.
</div>
<div id="ref-mehl2023fullstac" class="csl-entry" role="listitem">
Mehl, G.L., Seneviratne, M.G., Berg, M.L., Bidani, S., Distler, R.L., Gorgens, M., Kallander, K.E., Labrique, A.B., Landry, M.S., Leitner, C., Lubell-Doughtie, P.B., Marcelo, A.D., Matias, Y., Nelson, J., Nguyen, V., Nsengimana, J.P., Orton, M., Otzoy Garcia, D.R., Oyaole, D.R., Ratanaprayul, N., Roth, S., Schaefer, M.P., Settle, D., Tang, J., Tien-Wahser, B., Wanyee, S., Hersch, F., 2023. A full-<span>STAC</span> remedy for global digital health transformation: Open standards, technologies, architectures and content. Oxford Open Digital Health 1, oqad018. <a href="https://doi.org/10.1093/oodh/oqad018">https://doi.org/10.1093/oodh/oqad018</a>
</div>
<div id="ref-mullie2023coda" class="csl-entry" role="listitem">
Mullie, L., Afilalo, J., Archambault, P., Bouchakri, R., Brown, K., Buckeridge, D.L., Cavayas, Y.A., Turgeon, A.F., Martineau, D., Lamontagne, F., Lebrasseur, M., Lemieux, R., Li, J., Sauthier, M., St-Onge, P., Tang, A., Witteman, W., Chassé, M., 2023. <span>CODA</span>: An open-source platform for federated analysis and machine learning on distributed healthcare data. Journal of the American Medical Informatics Association ocad235. <a href="https://doi.org/10.1093/jamia/ocad235">https://doi.org/10.1093/jamia/ocad235</a>
</div>
<div id="ref-india2020national" class="csl-entry" role="listitem">
National <span>Digital Health Mission</span>, 2020. India National Health Authority.
</div>
<div id="ref-nsaghurwe2021one" class="csl-entry" role="listitem">
Nsaghurwe, A., Dwivedi, V., Ndesanjo, W., Bamsi, H., Busiga, M., Nyella, E., Massawe, J.V., Smith, D., Onyejekwe, K., Metzger, J., Taylor, P., 2021. One country’s journey to interoperability: <span>Tanzania</span>’s experience developing and implementing a national health information exchange. BMC Medical Informatics and Decision Making 21, 139. <a href="https://doi.org/10.1186/s12911-021-01499-6">https://doi.org/10.1186/s12911-021-01499-6</a>
</div>
<div id="ref-openhie" class="csl-entry" role="listitem">
<span>OpenHIE Framework</span> v5.0, 2022. OpenHIE.
</div>
<div id="ref-pedreira2023composable" class="csl-entry" role="listitem">
Pedreira, P., Erling, O., Karanasos, K., Schneider, S., McKinney, W., Valluri, S.R., Zait, M., Nadeau, J., 2023. The <span>Composable Data Management System Manifesto</span>. Proceedings of the VLDB Endowment 16, 2679–2685. <a href="https://doi.org/10.14778/3603581.3603604">https://doi.org/10.14778/3603581.3603604</a>
</div>
<div id="ref-peng2023etlprocess" class="csl-entry" role="listitem">
Peng, Y., Henke, E., Reinecke, I., Zoch, M., Sedlmayr, M., Bathelt, F., 2023. An <span class="nocase">ETL-process</span> design for data harmonization to participate in international research with <span>German</span> real-world data based on <span>FHIR</span> and <span>OMOP CDM</span>. International Journal of Medical Informatics 169, 104925. <a href="https://doi.org/10.1016/j.ijmedinf.2022.104925">https://doi.org/10.1016/j.ijmedinf.2022.104925</a>
</div>
<div id="ref-reynolds2011open" class="csl-entry" role="listitem">
Reynolds, C.J., Wyatt, J.C., 2011. Open <span>Source</span>, <span>Open Standards</span>, and <span>Health Care Information Systems</span>. Journal of Medical Internet Research 13, e1521. <a href="https://doi.org/10.2196/jmir.1521">https://doi.org/10.2196/jmir.1521</a>
</div>
<div id="ref-rieke2020future" class="csl-entry" role="listitem">
Rieke, N., Hancox, J., Li, W., Milletarì, F., Roth, H.R., Albarqouni, S., Bakas, S., Galtier, M.N., Landman, B.A., Maier-Hein, K., Ourselin, S., Sheller, M., Summers, R.M., Trask, A., Xu, D., Baust, M., Cardoso, M.J., 2020. The future of digital health with federated learning. npj Digital Medicine 3, 1–7. <a href="https://doi.org/10.1038/s41746-020-00323-1">https://doi.org/10.1038/s41746-020-00323-1</a>
</div>
<div id="ref-sinaci2024privacypreserving" class="csl-entry" role="listitem">
Sinaci, A.A., Gencturk, M., Alvarez-Romero, C., Laleci Erturkmen, G.B., Martinez-Garcia, A., Escalona-Cuaresma, M.J., Parra-Calderon, C.L., 2024. Privacy-preserving federated machine learning on <span>FAIR</span> health data: <span>A</span> real-world application. Computational and Structural Biotechnology Journal 24, 136–145. <a href="https://doi.org/10.1016/j.csbj.2024.02.014">https://doi.org/10.1016/j.csbj.2024.02.014</a>
</div>
<div id="ref-smits2022improved" class="csl-entry" role="listitem">
Smits, D., Van Beusekom, B., Martin, F., Veen, L., Geleijnse, G., Moncada-Torres, A., 2022. An <span>Improved Infrastructure</span> for <span>Privacy-Preserving Analysis</span> of <span>Patient Data</span>, in: Mantas, J., Gallos, P., Zoulias, E., Hasman, A., Househ, M.S., Diomidous, M., Liaskos, J., Charalampidou, M. (Eds.), Studies in <span>Health Technology</span> and <span>Informatics</span>. IOS Press. <a href="https://doi.org/10.3233/SHTI220682">https://doi.org/10.3233/SHTI220682</a>
</div>
<div id="ref-summitinstitutefordevelopment2023bunda" class="csl-entry" role="listitem">
Summit Institute for Development, 2023. <span>BUNDA App</span>. Summit Institute.
</div>
<div id="ref-syzdykova2017opensource" class="csl-entry" role="listitem">
Syzdykova, A., Malta, A., Zolfo, M., Diro, E., Oliveira, J.L., 2017. Open-<span>Source Electronic Health Record Systems</span> for <span>Low-Resource Settings</span>: <span>Systematic Review</span>. JMIR Medical Informatics 5, e44. <a href="https://doi.org/10.2196/medinform.8131">https://doi.org/10.2196/medinform.8131</a>
</div>
<div id="ref-teo2024federated" class="csl-entry" role="listitem">
Teo, Z.L., Jin, L., Liu, N., Li, S., Miao, D., Zhang, X., Ng, W.Y., Tan, T.F., Lee, D.M., Chua, K.J., Heng, J., Liu, Y., Goh, R.S.M., Ting, D.S.W., 2024. Federated machine learning in healthcare: <span>A</span> systematic review on clinical applications and technical architecture. Cell Reports Medicine 5, 101419. <a href="https://doi.org/10.1016/j.xcrm.2024.101419">https://doi.org/10.1016/j.xcrm.2024.101419</a>
</div>
<div id="ref-thaiya2021adoption" class="csl-entry" role="listitem">
Thaiya, M.S., Julia, K., Joram, M., Benard, M., Nambiro, D.A., 2021. Adoption of <span>ICT</span> to <span>Enhance Access</span> to <span>Healthcare</span> in <span>Kenya</span>.
</div>
<div id="ref-tilahun2023african" class="csl-entry" role="listitem">
Tilahun, B., Mamuye, A., Yilma, T., Shehata, Y., 2023. African <span>Union Health Information Exchange Guidelines</span> and <span>Standards</span>.
</div>
<div id="ref-tsafnat2024converge" class="csl-entry" role="listitem">
Tsafnat, G., Dunscombe, R., Gabriel, D., Grieve, G., Reich, C., 2024. Converge or collide? <span>Making</span> sense of a plethora of open data standards in healthcare: An editorial.
</div>
</div></section><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>See <a href="https://en.wikipedia.org/wiki/GSM#Initial_European_development">https://en.wikipedia.org/wiki/GSM#Initial_European_development</a>.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p>https://build.fhir.org/ig/FHIR/sql-on-fhir-v2/<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3"><p>FHIR: <a href="https://confluence.hl7.org/display/FHIR/Open+Source+Implementations">https://confluence.hl7.org/display/FHIR/Open+Source+Implementations</a>. OMOP: <a href="https://www.ohdsi.org/software-tools/">https://www.ohdsi.org/software-tools/</a>. OpenEHR: <a href="https://openehr.org/products_tools/platform/">https://openehr.org/products_tools/platform/</a><a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn4"><p>https://deltaplus.azdelta.be/az-delta30/zorginnovatie<a href="#fnref4" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn5"><p>https://omoponfhir.org<a href="#fnref5" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn6"><p>See <a href="https://arrow.apache.org">https://arrow.apache.org</a>, <a href="https://parquet.apache.org">https://parquet.apache.org</a> and <a href="https://iceberg.apache.org">https://iceberg.apache.org</a><a href="#fnref6" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn7"><p><a href="https://hapifhir.io">https://hapifhir.io</a><a href="#fnref7" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn8"><p><a href="https://getdbt.com">https://getdbt.com</a><a href="#fnref8" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn9"><p><a href="https://www.getdbt.com/">https://www.getdbt.com/</a><a href="#fnref9" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn10"><p><a href="https://jembi.gitbook.io/instant-v2">https://jembi.gitbook.io/instant-v2</a><a href="#fnref10" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>